---
title: "Creating figures for the paper"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: true
    code_folding: hide
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
params:
  sce: NULL
  cell_group_name: NULL
  infercnv_dir: NULL
  metadata: NULL
  cache_dir: NULL
  rerun: true
  max_prop_for_wt: 0.5
  min_prop_for_cnv: 0.8
---

```{r setup, include = F, message = F, warning = F, class.source = "fold-hide"}
# module load R/4.4.0 ; Rscript -e 'rmarkdown::render("reports/paper_figures.Rmd", output_dir = "out/paper_figures/")' # nolint: line_length_linter.

# libraries
library(magrittr)
library(ggplot2)
library(SummarizedExperiment)
library(SingleCellExperiment)

# chunk options
knitr::opts_knit$set(root.dir = "../")
knitr::opts_chunk$set(warning = FALSE, dpi = 300, message = FALSE)
```

# Datasets

## LCM NanoSeq

```{r}
# dirs
lh22_dir <- "/lustre/scratch125/casm/team268im/lh22/Rheumatoid_Arthritis_Paper/"
nb15_dir <- "/lustre/scratch126/casm/team154pc/nb15/liver/"

# functions and libraries
source("bin/lolliplot.R")
library(magrittr)
library(ggplot2)

# load the annotated mutations from nanoseq output
load(file.path(lh22_dir, "Figures/Figure_2/LCM_NanoSeq_Overlap/GOI",
               "pileup_muts.rda"))
nano_muts <- nano_muts %>% tibble::as_tibble()

# run dNdScv
dndscv_in <- nano_muts %>% dplyr::select(sampleID, chr, pos, ref, mut)
dndscv_out <- dndscv::dndscv(dndscv_in)

# load refcds, add gene names
data("refcds_hg19", package = "dndscv")
names(RefCDS) <- RefCDS %>% purrr::map(~ .$gene_name)

# combine annotated mutations with dnds output
nanoseq_data <-
  nano_muts %>%
  dplyr::mutate(impact = dplyr::case_when(impact == "no-SNV" ~ "Indel",
                                          is.na(impact) ~ "Non-coding",
                                          TRUE ~ impact)) %>%
  dplyr::select(sampleID, chr, pos, ref, mut, duplex_cov, impact, donor) %>%
  dplyr::left_join(dndscv_out$annotmuts) %>%
  # rename cols
  dplyr::rename(CHROM = chr, POS = pos, REF = ref, ALT = mut)

# plotting params
filterNonCoding <- TRUE
pointWidth <- 0.017

# define genes of interest
g <- c("NF1")

# get gene info
g_CDS <- RefCDS[[g]]

# Â get gene intervals
intervals_cds0 <- tibble::tibble(
  chr = g_CDS$chr,
  strand = g_CDS$strand,
  start = g_CDS$intervals_cds[, 1],
  end = g_CDS$intervals_cds[, 2]
) %>%
  dplyr::arrange(start) %>%
  dplyr::mutate(length = (end + 1) - start)

# get genomic position to CDS position index
cds_coords_index <- getCDSindex(intervals_cds0)

# convert CDS intervals
intervals_cds <-
  intervals_cds0 %>%
  dplyr::left_join(cds_coords_index %>%
                     dplyr::select(start = pos, cdsStart = cdsPos),
                   by = "start") %>%
  dplyr::left_join(cds_coords_index %>%
                     dplyr::select(end = pos, cdsEnd = cdsPos),
                   by = "end")

# subset nanoseq_data to variants in the gene
gene0 <-
  nanoseq_data %>%
  dplyr::filter(CHROM == unique(intervals_cds$chr),
                POS >= min(intervals_cds$start),
                POS <= max(intervals_cds$end))

# get coords outside the CDS
noncds <-
  cds_coords_index %>%
  dplyr::rename(POS = pos) %>%
  dplyr::bind_rows(getCDSindexNonCDS(cds_coords_index, gene0,
                                     strand = unique(intervals_cds$strand)))

gene <-
  gene0 %>%
  dplyr::left_join(noncds, by = "POS") %>%
  dplyr::arrange(cdsPos) %>%
  tidyr::drop_na(cdsPos)

# set impact colours
impact_palette <-
  c("Non-coding" = "gray100",
    "Synonymous" = "gray76",
    "Missense" = "#abd9e9",
    "Nonsense" = "#fdae61",
    "Indel" = "pink",
    "Essential_Splice" = "#fee090",
    "Stop_loss" = "#bb0f12")
gene <-
  gene %>%
  dplyr::mutate(impact = factor(impact, levels = names(impact_palette)),
                col = impact_palette[impact])

# set label positions
lab_pos <-
  jitterLabels(sort(gene$cdsPos),
               xscale = c(1, max(cds_coords_index$cdsPos)),
               lineW = pointWidth) %>%
  reAdjustLabels(lineW = pointWidth, xmax = max(cds_coords_index$cdsPos))
variants <-
  gene %>%
  dplyr::mutate(labPos = lab_pos,
                aachange = tidyr::replace_na(aachange, "."))

# xmax and xmin
x_max <-
  max(c(intervals_cds$cdsEnd, intervals_cds$cdsStart)) +
  max(c(intervals_cds$cdsEnd, intervals_cds$cdsStart)) * 0.05
x_min <- 1 - x_max * 0.01

# plot variants
p_var <-
  variants %>%
  dplyr::group_by(CHROM, POS, REF, ALT) %>%
  dplyr::mutate(n_donors = dplyr::n_distinct(donor)) %>%
  ggplot() +
  # variant stems
  geom_segment(aes(x = cdsPos, xend = cdsPos, y = 0,
                   yend = (0.125 * max(lab_pos)))) +
  geom_segment(aes(x = cdsPos, xend = labPos, y = (0.125 * max(lab_pos)),
                   yend = (0.25 * max(lab_pos)))) +
  geom_segment(aes(x = labPos, xend = labPos, y = (0.25 * max(lab_pos)),
                   yend = (0.325 * max(lab_pos)))) +
  # variant points
  geom_point(aes(x = labPos, y = 0.325 * max(lab_pos), fill = impact,
                 size = n_donors),
             shape = 21, colour = "black", size = 5) +
  # label aa change
  # geom_text(aes(label = name, x = labPos, y = (0.5 * max(lab_pos))),
  #           angle = 90) +
  theme_void() +
  scale_fill_manual(guide = "legend", values = impact_palette) +
  scale_x_continuous(expand = c(0, 0), limits = c(x_min, x_max)) +
  coord_cartesian(clip = "off") +
  scale_y_continuous(limits = c(0, max(lab_pos)), expand = c(0, 0)) +
  ggtitle(g)

# plot CDS
p_cds <-
  intervals_cds %>%
  ggplot(aes(xmin = cdsStart, xmax = cdsEnd, ymin = 0, ymax = 1)) +
  geom_rect(fill = "#d0d1e6", colour = "black") +
  theme_void() + 
  xlab("CDS") +
  scale_x_continuous(expand = c(0, 0), limits = c(x_min, x_max)) +
  scale_y_continuous(expand = c(0, 0)) +
  theme(plot.margin = unit(c(0, 0.5, 0.5, 0.5), "cm"))

pdf("test.pdf", width = 20)
cowplot::plot_grid(p_var, p_cds, ncol = 1, rel_heights = c(15, 1))
dev.off()

```

# Loading the inferCNV outputs

inferCNV was run on the scRNA-seq data to call copy number variants in synovial 
cells. No reference group was used, so the average profile of all cells was used 
in each patient as the baseline. 

First, we load the CITE-seq data as a `SingleCellExperiment` object. The CNV 
metadata from the `inferCNV` run has been aded to this object.

```{r load_sce, results = "asis", class.source = "fold-show"}
# load sce with gains
sce <-
  "out/Zhang2023/by_fibroblast_cluster/summary/gains_and_losses_cache/" %>%
  list.files(pattern = "sce_w_gains", full.names = TRUE) %>%
  readRDS()
```

# Quantifying CNVs of interest in the samples' fibroblast populations

We first prepare the data so that we can view the proportions of chr7 and chr5 
gains in each patient's fibroblast population.

```{r}
events <-
  sce@colData %>%
  tibble::as_tibble() %>%
  dplyr::select(sample, cell, cell_group, chr7_gain, chr5_gain) %>%
  tidyr::pivot_longer(-c("sample", "cell", "cell_group")) %>%
  dplyr::group_by(sample, cell, cell_group) %>%
  # remove cells in which no calls made
  dplyr::filter(!all(is.na(value))) %>%
  dplyr::mutate(event = dplyr::case_when(all(is.na(value) | value == 0) ~ "WT",
                                         value == 1 ~ name, TRUE ~ NA)) %>%
  dplyr::summarise(
    events = paste(sort(unique(na.omit(event))), collapse = ",") %>%
      factor(levels = c("chr5_gain,chr7_gain", "chr7_gain", "chr5_gain", "WT")),
    n_events = sum(value, na.rm = TRUE)) %>%
  # get props and counts
  dplyr::ungroup() %>%
  dplyr::add_count(sample, name = "n_cells_per_sample") %>%
  dplyr::add_count(sample, events, name = "n_cells_per_sample_x_event") %>%
  dplyr::distinct(sample, events,
                  n_cells_per_sample_x_event, n_cells_per_sample) %>%
  dplyr::mutate(prop_cells_per_sample_x_event = n_cells_per_sample_x_event /
                  n_cells_per_sample) %>%
  # refactor samples by prop non-WT cells
  dplyr::group_by(sample) %>%
  dplyr::mutate(
    n_wt_per_sample = dplyr::case_when(events == "WT" ~
                                         n_cells_per_sample_x_event,
                                       TRUE ~ 0),
    prop_non_wt_per_sample = (n_cells_per_sample - sum(n_wt_per_sample)) /
                  n_cells_per_sample) %>%
  dplyr::ungroup() %>%
  dplyr::mutate(sample = forcats::fct_reorder2(sample, n_cells_per_sample,
                                               prop_non_wt_per_sample)) %>%
  tidyr::pivot_longer(c("n_cells_per_sample_x_event",
                        "prop_cells_per_sample_x_event"))
```

# Filtering out patients with few fibroblasts

```{r}
min_n_fibroblasts_per_sample <- 100
```

We set the minimum number of fibroblasts per patient to 
`r min_n_fibroblasts_per_sample`. We can review how many patients pass the 
cut-off to give us reliable results.

```{r plot_fibroblast_dist}
n_fibroblasts <-
  sce@colData %>%
  tibble::as_tibble() %>%
  dplyr::count(sample)
n_fibroblasts %>%
  ggplot(aes(x = n)) +
  geom_histogram(binwidth = 20) +
  geom_vline(xintercept = min_n_fibroblasts_per_sample, linetype = "dashed", 
             colour = "red") +
  theme_classic()
```

```{r cut_off}
enough_fibroblasts <-
  dplyr::filter(n_fibroblasts, n >= min_n_fibroblasts_per_sample)$sample
n_fibroblasts %>%
  dplyr::summarise(`n >= 100` = sum(n >= 100),
                   `n < 100` = sum(n < 100),
                   n = dplyr::n()) %>%
  knitr::kable()
```

This leaves `r length(enough_fibroblasts)` / `r nrow(n_fibroblasts)` of the 
samples. We will filter the `SingleCellExperiment` object accordingly.

We can also review the relationship between the number of cells in each 
patient's fibroblast population and the proportions / counts of the CNVs of 
interest (chr7 gain and chr5 gain).

```{r plot_event_vs_fibroblast}
events %>%
  ggplot(aes(x = sample, y = value, fill = events)) +
  geom_col() +
  scale_x_discrete(guide = guide_axis(angle = -90)) +
  facet_wrap(~ name, scale = "free_y", ncol = 1) +
  theme_classic() +
  scale_fill_brewer(palette = "Dark2") +
  theme(legend.position = "bottom")
```

We can now apply the cut-off.

```{r apply_cutoff}
sce <- sce[, sce$sample %in% enough_fibroblasts]
events <- events %>% dplyr::filter(sample %in% enough_fibroblasts)
```

After applying the cut-off, we look for sample-size effects once again.

```{r plot_event_vs_fibroblast_after_cutoff}
events %>%
  ggplot(aes(x = sample, y = value, fill = events)) +
  geom_col() +
  scale_x_discrete(guide = guide_axis(angle = -90)) +
  facet_wrap(~ name, scale = "free_y", ncol = 1) +
  theme_classic() +
  scale_fill_brewer(palette = "Dark2") +
  theme(legend.position = "bottom")
```

We can see that there is still a strong relationship between the size of the
fibroblast populations by patient and the proportion of those cells that are
called as having a chr7 or chr5 gain. This is most likely related to with the
fact that we set `ref_group_names = NULL`. This compounds the effect of the
number of cells on the ability to detect CNVs (the false negative rate), 
because...

1. having fewer cells makes it less likely to capture these somewhat rare events

2. having fewer cells makes the reference profile noisier and less
representative of the 'baseline' expression levels across chromosomes, further
limiting CNV calling

Keeping this in mind, we will proceed.

# Plot 1 - chr7, chr5, and chr3 barplots by sample

```{r}
p <-
  sce@colData %>%
  tibble::as_tibble() %>%
  dplyr::add_count(sample) %>%
  tidyr::pivot_longer(cols = c("chr7_gain", "chr5_gain", "chr3_gain")) %>%
  dplyr::filter(!is.na(value)) %>%
  dplyr::mutate(value = as.character(value)) %>%
  ggplot(aes(x = reorder(sample, -n), fill = value)) +
  scale_x_discrete(guide = guide_axis(angle = -90)) +
  theme_classic() +
  scale_fill_brewer(palette = "Dark2") +
  facet_grid(name ~ .)
p + geom_bar() + labs(x = "sample", y = "n cells")
p + geom_bar(position = "fill") + labs(x = "sample", y = "proportion of cells")
```

# Plot 2 - chr7, chr5, and chr3 barplots by celltype

```{r}
cnvs <-
  "out/Zhang2023/by_celltype/summary/gains_and_losses_cache/" %>%
  list.files(pattern = "cnvs", full.names = TRUE) %>%
  readRDS()
p <-
  cnvs %>%
  dplyr::filter(event == "dupli", measure == "proportion",
                (value < 0.5 | value > 0.8),
                chr %in% c("chr7", "chr5", "chr3")) %>%
  dplyr::mutate(
    value = dplyr::case_when(value <= 0.5 ~ 0, value >= 0.8 ~ 1) %>%
      as.character()) %>%
  ggplot(aes(x = cell_type, fill = value)) +
  theme_classic() +
  facet_grid(chr ~ .) +
  scale_fill_brewer(palette = "Dark2") +
  scale_x_discrete(guide = guide_axis(angle = -45))
p + geom_bar()
p + geom_bar(position = "fill") + ylab("proportion")
```

# Plot 3 - number of transcripts per cell vs CNVs

Next, we look at the distribution of the the number of transcripts in cells in 
which chr7 gain has been detected versus those without. If the distribution is 
greatly skewed in one or the other, this would suggest that this technical 
variate is strongly affecting the performance of `inferCNV` between cells.

```{r}
colSums(sce@assays@data$counts) %>%
  tibble::enframe(name = "cell", value = "n_transcripts") %>%
  dplyr::inner_join(sce@colData %>% tibble::as_tibble()) %>%
  dplyr::filter(!is.na(chr7_gain)) %>%
  dplyr::mutate(chr7_gain = as.character(chr7_gain)) %>%
  ggplot(aes(x = n_transcripts, fill = chr7_gain)) +
  geom_histogram(alpha = 0.5) +
  theme_classic()
```

The distributions appear to be sensible. 

# Plot 4 - inferCNV plot with all individuals together

I am currently trying to run this. I have run into some issues with `inferCNV`
dependencies breaking on farm22, so this has been on hold for the last 2 weeks.

# Plot 5 - Mutation hotspots

We load Luke's object. The `nano_muts` are the targeted mutations. The 
`exome_muts` is a list of significant drivers in each cell type (dNdScv output).


```{r}
# add library path
.libPaths(c("/lustre/scratch125/casm/team268im/at31/.R/lib/4.2.2/", .libPaths()))

# libraries
library(magrittr)
library(GenomicRanges)
library(trackViewer)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)  # package for gene model
library(org.Hs.eg.db)                       # package for gene name
library(biomaRt)
library(httr)                               # package for REST API

# dirs
lh22_dir <- "/lustre/scratch125/casm/team268im/lh22/Rheumatoid_Arthritis_Paper/"

# load the metadata
sm <- readr::read_tsv(file.path(lh22_dir, "Data/MetaData/sample_meta.tsv"))

# load the muts (targeted mutations in the nano_muts object)
load(file.path(lh22_dir, "Figures/Figure_2/LCM_NanoSeq_Overlap/GOI",
               "pileup_muts.rda"))
nano_muts <-
  nano_muts %>%
  tibble::as_tibble()

# get mutations, convert to granges
gr_muts_ls <-
  nano_muts %>%
  tibble::as_tibble() %>%
  dplyr::add_count(mut_site) %>%
  {
    GenomicRanges::GRanges(
      seqnames = paste0("chr", .$chr),
      ranges = IRanges::IRanges(start = .$pos,
                                width = 1,
                                names = .$mut_site),
      strand = ifelse(.$strand == 1, "+", "-"),
      gene = .$gene, score = .$n)
  } %>%
  {split(., .$gene)}

# get the refcds
data("refcds_hg19", package = "dndscv")
names(RefCDS) <- RefCDS %>% purrr::map(~ .$gene_name)

# write urls for proteinpaint

# write custom input file for proteinpaint
nano_muts %>%
  dplyr::filter(gene == "NF1") %>%
  dplyr::transmute(
    aachange, 
    aapos = as.numeric(stringr::str_sub(aachange, 2, -2)),
    impact = dplyr::case_when(impact == "Missense" ~ "M",
                              impact == "Nonsense" ~ "N",
                              impact == "Synonymous" ~ "S",
                              impact == "no-SNV" & type == "del" ~ "D",
                              impact == "no-SNV" & type == "ins" ~ "I",
                              TRUE ~ NA),
    donor) %>%
  dplyr::filter(!is.na(aapos)) %>%
  readr::write_csv("protein_paint.csv")

# write custom input file for mutationmapper
nano_muts %>%
  dplyr::filter(gene == "NF1") %>%
  dplyr::transmute(
    Chromosome = chr, Start_Position = pos - 1, End_Position = pos,
    Reference_Allele = ref, Variant_Allele = mut, Hugo_Symbol = gene
  ) %>%
  readr::write_tsv("mutation_mapper.tsv")

# get the gene
g <- "NF1"

# retrieve refcds element for gene
x <- RefCDS[[g]]

# get the position of each base along the exons
# RefCDS is 1-based - must convert to 0-based for calculations, then
# convert back to 1-based for plotting
exons <-
  tibble::tibble(start = x$intervals_cds[, 1] - 1,
                 end = x$intervals_cds[, 2]) %>%
  dplyr::mutate(dplyr::across(everything(), 
                              ~ .x - (x$intervals_cds[1, 1] - 1),
                              .names = "rel_{.col}"),
                exon = dplyr::row_number(),
                exon_length = end - start,
                cds_end = cumsum(exon_length),
                cds_start = dplyr::lag(cds_end, default = 0),
                # convert starts to 1-based for plotting
                dplyr::across(dplyr::contains("start"),
                              ~ .x + 1))

# expand to get every position
exons_exp <-
  exons %>%
  dplyr::group_by(rel_start, rel_end, exon, exon_length) %>%
  dplyr::reframe(pos = start:end, cds_pos = cds_start:cds_end)

# load interpro entries
interpro2go <-
  readLines("data/interpro/interpro2go.txt") %>%
  tibble::as_tibble() %>%
  dplyr::filter(!grepl("^!", value)) %>%
  tidyr::separate_wider_delim("value", delim = " ",
                              names = c("interpro_accession", "rest"),
                              too_many = "merge") %>%
  tidyr::separate(rest, sep = " > GO:| ; GO:",
                  into = c("domain_name", "GO_name", "GO_id")) %>%
  dplyr::mutate(interpro_accession = gsub("InterPro:", "", interpro_accession))
interpro <-
  readr::read_tsv("data/interpro/entry.list") %>%
  dplyr::rename(interpro_accession = ENTRY_AC)

# retrieve protein domain annotations
library(ensembldb)
library(EnsDb.Hsapiens.v75)
edb <- EnsDb.Hsapiens.v75
domains <-
  transcripts(edb, filter = GeneNameFilter(g),
              columns = c("protein_id", "tx_biotype", "uniprot_id",
                          "protein_domain_id", "protein_domain_source",
                          "prot_dom_start", "prot_dom_end",
                          "interpro_accession"))
domains_exp <-
  txs %>%
  dplyr::as_tibble() %>%
  dplyr::left_join(interpro2go) %>%
  dplyr::filter(tx_biotype == "protein_coding",
                !is.na(interpro_accession), !is.na(domain_name)) %>%
  dplyr::mutate(dom_start = start + prot_dom_start,
                dom_end = start + prot_dom_end) %>%
  dplyr::group_by(tx_id, protein_id, protein_domain_id, protein_domain_source,
                  prot_dom_start, prot_dom_end, interpro_accession,
                  domain_name, GO_name) %>%
  dplyr::reframe(pos = dom_start:dom_end)

# combine all exon and domain annotations
annots <-
  exons_exp %>%
  dplyr::full_join(domains_exp)

# add exon annots to muts
impact_palette <-
  c("Synonymous" = "#b3b3b3",
    "Missense" = "#5f9ea0",
    "Nonsense" = "#68218a")
p_dat <-
  nano_muts %>%
  dplyr::group_by(chr, pos, impact) %>%
  dplyr::count() %>%
  dplyr::right_join(annots) %>%
  dplyr::mutate(alternate = ifelse(exon %% 2 == 0, "e", "o"),
                impact = factor(impact, levels = names(impact_palette)))

# plot
pdf("test.pdf", width = 30)
# mutations
p_dat %>%
  ggplot(aes(x = cds_pos, y = n, fill = impact, colour = impact)) +
  geom_col(width = 1) +
  geom_point() +
  theme_classic() +
  scale_fill_manual(values = impact_palette,
                    na.value = "white", na.translate = FALSE) +
  scale_colour_manual(values = impact_palette,
                      na.value = "white", na.translate = FALSE)
# exon bounds
p_dat %>%
  dplyr::group_by(exon, alternate) %>%
  dplyr::summarise(xmin = min(cds_pos), xmax = max(cds_pos)) %>%
  ggplot(aes(xmin = xmin, xmax = xmax, ymin = -1, ymax = 1,
             fill = alternate)) +
  geom_hline(yintercept = 0, colour = "black") +
  geom_rect(colour = "black") +
  theme_void() +
  theme(axis.text.x = element_text()) +
  scale_fill_manual(values = c("e" = "#e3b3b2",
                               "o" = "#cd6f6e"))
# protein domains
p_dat %>%
  dplyr::filter(!is.na(protein_domain_id)) %>%
  dplyr::group_by(tx_id, domain_name, GO_name, protein_domain_id) %>%
  dplyr::summarise(xmin = min(cds_pos), xmax = max(cds_pos)) %>%
  ggplot(aes(xmin = xmin, xmax = xmax, ymin = -1, ymax = 1,
             fill = GO_name)) +
  geom_hline(yintercept = 0, colour = "black") +
  geom_rect(colour = "black") +
  theme_void() +
  scale_fill_discrete(na.value = "white", na.translate = FALSE) +
  facet_grid(tx_id ~ .)

dev.off()



# get gene coords (GRCh37)
ensembl <- useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl",
                      GRCh = 37)
bm <-
  getBM(attributes = c("chromosome_name", "start_position",
                       "end_position", "strand", "hgnc_symbol"),
        filters = c("hgnc_symbol"),
        values = list(names(gr_muts_ls)),
        mart = ensembl)
gr_genes <-
  GenomicRanges::GRanges(
    seqnames = paste0("chr", bm$chromosome_name),
    ranges = IRanges::IRanges(start = bm$start_position,
                              end = bm$end_position,
                              names = bm$hgnc_symbol),
    strand = ifelse(bm$strand == 1, "+", "-"),
    gene = bm$hgnc_symbol)
gr_genes_ls <- split(gr_genes, gr_genes$gene)

# get transcripts per gene
nf1_trs <-
  geneModelFromTxdb(TxDb.Hsapiens.UCSC.hg19.knownGene,
                    org.Hs.eg.db, gr = gr_genes_ls$NF1) %>%
  purrr::map(~ .$dat) %>%
  {Reduce(c, .)} %>%
  subset(symbol == "NF1" & feature == "CDS")
nf1_trs_fill <-
  tibble::tibble(transcript = unique(nf1_trs$transcript),
                 fill = rainbow(dplyr::n_distinct(nf1_trs$transcript))) %>%
  tibble::column_to_rownames("transcript")
nf1_trs$fill <-  nf1_trs_fill[nf1_trs$transcript, ]
nf1_trs$featureLayerID <- nf1_trs$transcript
names(nf1_trs) <- nf1_trs$transcript

# get protein domains
APIurl <- "https://www.ebi.ac.uk/proteins/api/" # base URL of the API
taxid <- "9606" # human tax ID
gene <- "NF1" # target gene
orgDB <- "org.Hs.eg.db" # org database to get the uniprot accession id
eid <- mget(gene, get(sub(".db", "SYMBOL2EG", orgDB)))[[1]]
chr <- mget(eid, get(sub(".db", "CHR", orgDB)))[[1]]
accession <- unlist(lapply(eid, function(.ele){
  mget(.ele, get(sub(".db", "UNIPROT", orgDB)))
  }))
featureURL <- paste0(APIurl,
                      "features?offset=0&size=-1&reviewed=true",
                      "&types=DNA_BIND%2CMOTIF%2CDOMAIN",
                      "&taxid=", taxid,
                      "&accession=", paste(accession, collapse = "%2C"))
response <- GET(featureURL)
content <- httr::content(response)[[1]]
acc <- content$accession
sequence <- content$sequence
gr <- GRanges(chr, IRanges(1, nchar(sequence)))
domains <- do.call(rbind, content$features)
domains <- GRanges(chr, IRanges(as.numeric(domains[, "begin"]),
                                as.numeric(domains[, "end"]),
                                names = domains[, "description"]))
names(domains)[1] <- "DNA_BIND" ## this is hard coding.
domains$fill <- 1 + seq_along(domains)
domains$height <- 0.04
## GET variations. This part can be replaced by user-defined data.
variationURL <- paste0(APIurl,
                        "variation?offset=0&size=-1",
                        "&sourcetype=uniprot&dbtype=dbSNP",
                        "&taxid=", taxid,
                        "&accession=", acc)
response <- GET(variationURL)
content <- httr::content(response)[[1]]
keep <- sapply(content$features, function(.ele) length(.ele$evidences)>2 && # filter the data by at least 2 evidences
                  !grepl("Unclassified", .ele$clinicalSignificances)) # filter the data by classified clinical significances.
nkeep <- c("wildType", "alternativeSequence", "begin", "end",
            "somaticStatus", "consequenceType", "score")
content$features <- lapply(content$features[keep], function(.ele){
  .ele$score <- length(.ele$evidences)
  unlist(.ele[nkeep]) 
})
variation <- do.call(rbind, content$features)
variation <-
  GRanges(chr,
          IRanges(as.numeric(variation[, "begin"]),
                  width = 1,
                  names = paste0(variation[, "wildType"],
                                  variation[, "begin"],
                                  variation[, "alternativeSequence"])),
          score = as.numeric(variation[, "score"]),
          color = as.numeric(factor(variation[, "consequenceType"]))+1)
variation$label.parameter.gp <- gpar(cex=.5)
seqlevels(variation) <- paste0("chr", seqlevels(variation))
seqlevels(domains) <- paste0("chr", seqlevels(domains))

domains %>%
  tibble::as_tibble()

pdf("test.pdf")
lolliplot(gr_muts_ls$NF1, domains, nf1_trs)
dev.off()

# plot transcripts per gene
pdf("test.pdf", width = 20)
lolliplot(gr_muts_ls$NF1, nf1_trs)
dev.off()
```

```{r}
load(file="/lustre/scratch125/casm/team268im/lh22/Rheumatoid_Arthritis_Paper/Figures/Figure_1/GOI.rda")
goi <- list(exome_genes = exome_genes, goi_exome = goi_exome, 
            goi_targeted = goi_targeted, targeted_genes = targeted_genes)

nanoseqVarsPath <- "nano_muts.tsv"
```