---
title: "Mutation loads in muscle ageing data from SComatic"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: true
    code_folding: show
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
params:
    do_rerun : TRUE
    
---

```{css style settings, echo = FALSE}
blockquote {
    padding: 10px 20px;
    margin: 0 0 20px;
    font-size: 14px;
    border-left: 5px solid #eee;
}
```

```{r setup, include = F, message = F, warning = F, class.source = 'fold-hide'}
# Rscript -e "rmarkdown::render('reports/mutation_load.Rmd', output_dir = 'out/mutation_load/')"
knitr::opts_knit$set(root.dir = '../')
knitr::opts_chunk$set(
  warning = FALSE, 
  dpi = 300, 
  message = FALSE,
  cache.path = 'reports/mutation_load_cache/') 

# libraries
library(magrittr)
library(ggplot2)
library(gridExtra)
```

```{r plotting_fns, include = F, echo = F}
get_hq_cts <- 
  function(ct_info, min_n_cells_per_celltype_across_ids) {
    ct_info %>% 
      dplyr::group_by(id, lvl, annotation_level1_alias) %>% 
      dplyr::filter(min(n_cells_per_id_per_celltype) >= min_n_cells_per_celltype_across_ids) %>%
      dplyr::ungroup() %>%
      dplyr::select(id, lvl, annotation_level1_alias)
  }

colour_kbl_column <- function(kbl, df, col) {
  kbl %>% 
    kableExtra::column_spec(
      which(colnames(df) == col), 
    color = "white",
    background = kableExtra::spec_color(df[, col, drop = T]))
}

plot_vafs_density <-
  function(som_muts) {
    som_muts %>%
      dplyr::filter(lvl == 'donor_id') %>%
      {split(., .$lvl)} %>%
      purrr::map(function(df) {
       df %>% 
          dplyr::group_by(lvl, id, annotation_level1) %>%
          dplyr::mutate(n = dplyr::n()) %>%
          ggplot(aes(x = VAF, label = n, fill = age_bracket)) +
          geom_density() +
          geom_vline(aes(xintercept = median_VAF_per_id_per_celltype), colour = 'grey') +
          geom_text(aes(x = 0.6, y = Inf), vjust = 1.3) +
          ggh4x::facet_nested(lvl + age_bracket + id ~ annotation_level1, scales = 'free_y') +
          theme_bw() +
          theme(axis.text.x = element_text(angle = -90, hjust = 0),
                axis.text.y = element_blank(),
                strip.text.y.right = element_text(angle = 0, hjust = 0),
                panel.grid.major = element_blank(),
                panel.grid.minor = element_blank())
      }) %>%
      {do.call("grid.arrange", c(., ncol = length(.)))}
  }

plot_vafs_boxplot <-
  function(som_muts) {
    som_muts %>%
      {split(., .$lvl)} %>%
      purrr::map(function(df) {
       df %>% 
          ggplot(aes(x = VAF)) +
          geom_boxplot() +
          ggh4x::facet_nested(annotation_level1_alias ~ lvl + age_bracket + donor_id + id, scales = 'free_y') +
          theme_bw() +
          theme(axis.text.y = element_blank(),
                axis.text.x = element_text(angle = -90, hjust = 0),
                strip.text.y.right = element_text(angle = 0, hjust = 0),
                panel.grid.major = element_blank(),
                panel.grid.minor = element_blank())
      }) %>%
      {do.call("grid.arrange", c(., ncol = length(unique(som_muts$lvl))))}
  }

plot_ct_vs_mut_load_per_id_bar <- 
  function(som_mut_loads) {
    som_mut_loads %>%
    {split(., .$lvl)} %>%
      purrr::map(function(lvl_i) {
        lvl_i %>%
          ggplot(aes(x = annotation_level1, y = som_mut_load_per_Mb,
                     fill = age_bracket)) +
          geom_col() +
          theme_bw() +
          theme(axis.text.x = element_text(angle = 270, hjust = 0, vjust = 0.5, size = 5),
                legend.position = 'none') +
          ggh4x::facet_nested(. ~ lvl + age_bracket + donor_id + id) 
      }) %>%
    {do.call("grid.arrange", c(., ncol = 1))}
  }

plot_ct_vs_mut_load_per_id_bar2 <- 
  function(som_mut_loads) {
    som_mut_loads %>%
    {split(., .$lvl)} %>%
      purrr::map(function(lvl_i) {
        lvl_i %>%
          ggplot(aes(x = annotation_level1, y = som_mut_load_per_Mb, 
                     fill = age_bracket)) +
          geom_col(position = position_dodge2(preserve = "single")) +
          theme_minimal() +
          theme(axis.text.x = element_text(angle = -45, hjust = 0, vjust = 0.5, size = 10)) 
      }) %>%
    {do.call("grid.arrange", c(., ncol = 1))}
  }

plot_celltype_vs_mut_load_per_id_heatmap <- 
  function(som_mut_loads) {
    pre_mat <-
      som_mut_loads %>%
      dplyr::ungroup() %>%
      dplyr::select(som_mut_load_per_Mb, annotation_level1, id, age_bracket, lvl) %>%
      tidyr::pivot_wider(names_from = annotation_level1, values_from = som_mut_load_per_Mb) %>%
      tibble::column_to_rownames('id') 
    mat <- 
      pre_mat %>%
      dplyr::select(-c(age_bracket, lvl)) %>%
      as.matrix()
    annot <-
      pre_mat[, c('lvl', 'age_bracket')]
    pheatmap::pheatmap(
      mat, cluster_rows = F, cluster_cols = F,
      annotation_row = annot)
  }

plot_celltype_vs_mut_load_vs_n_cells_per_id_heatmap <-
  function(som_mut_loads) {
    som_mut_loads %>%
      ggplot(aes(x = id, y = annotation_level1, 
                 colour = som_mut_load_per_Mb, 
                 size = n_cells_per_id_per_celltype)) +
      geom_point() +
      theme_bw() +
      viridis::scale_color_viridis() +
      theme(axis.text.x = element_text(angle = -90, hjust = 0),
            panel.grid = element_blank()) +
      ggh4x::facet_nested(~ lvl + age_bracket, 
                          scales = 'free_x', space = 'free_x')
  }

plot_mut_load_btwn_replicates <- 
  function(som_mut_loads) {
    p_dat <-
      som_mut_loads %>%
      dplyr::filter(lvl == 'sample_id') %>%
      dplyr::group_by(donor_id, age_bracket) %>%
      dplyr::filter(dplyr::n_distinct(id) == 2) %>%
      dplyr::mutate(rep = paste0('rep', as.numeric(as.factor(id)))) %>%
      tidyr::pivot_wider(id_cols = c('donor_id', 'age_bracket', 'annotation_level1'),
                         names_from = rep, values_from = som_mut_load_per_Mb) 
    if (nrow(p_dat) > 0) {
      p_dat %>%
      #dplyr::filter(!is.na(rep1), !is.na(rep2)) %>%
      ggplot(aes(x = rep1, y = rep2, colour = annotation_level1, group = 1)) +
      geom_smooth(method = 'lm') +
      geom_point() + 
      ggh4x::facet_nested(age_bracket + donor_id ~ .) +
      theme_bw() +
      coord_equal()
    }
  }

plot_n_cells_vs_metrics <- 
  function(som_mut_loads) {
    som_mut_loads %>% 
      tidyr::pivot_longer(cols = c('n_callable_sites', 'n_som_muts', 'som_mut_load_per_Mb')) %>%
      ggplot(aes(x = n_cells_per_id_per_celltype, y = value,
                 colour = donor_id, shape = age_bracket)) +
      geom_point() +
      theme_bw() +
      theme(aspect.ratio = 1) +
      facet_wrap(lvl ~ name, scales = 'free_y') 
  }

plot_mutational_signatures <- function(som_muts) {
  # get all possible trios
  all_REF_trios <-
    c('A', 'T', 'G', 'C') %>%
    {tibble::tibble(up = ., REF_fix = ., down = ., ALT_fix = .)} %>%
    tidyr::expand(up, REF_fix, down, ALT_fix) %>%
    dplyr::filter(REF_fix %in% c('C', 'T'), ALT_fix != REF_fix) %>%
    dplyr::transmute(ALT_fix, REF_fix,
                     REF_trio = paste0(up, REF_fix, down),
                     REF_label = paste0(up, '**', REF_fix, '**', down)) %>%
    dplyr::cross_join(
      som_muts %>% 
        dplyr::ungroup() %>%
        dplyr::distinct(lvl, id, age_bracket) 
    )
  
  # mutation colours
  mut_cols <-
    c('C > A' = '#1493d2',
      'C > G' = '#070707',
      'C > T' = '#bb0f12',
      'T > A' = '#ababa5',
      'T > C' = '#6ab538',
      'T > G' = '#dfa99a')
  
  # plot
  som_muts %>%
    dplyr::filter(lvl == 'donor_id') %>%
    dplyr::mutate(dplyr::across(
      c('REF', 'ALT', 'Up_context', 'Down_context'), 
      ~ dplyr::case_when(REF %in% c('G', 'A') ~ chartr("ATGC", "TACG", .x),
                         TRUE ~ .x), 
      .names = "{.col}_fix"),
      REF_trio = paste0(substr(Up_context_fix, 5, 5), REF_fix, substr(Down_context_fix, 1, 1))) %>%
    dplyr::group_by(REF_trio, REF_fix, ALT_fix, id, age_bracket) %>%
    dplyr::count() %>%
    dplyr::group_by(id) %>%
    dplyr::mutate(prop = n / sum(n)) %>%
    dplyr::full_join(all_REF_trios) %>%
    dplyr::mutate(mutation = paste(REF_fix, '>', ALT_fix),
                  dplyr::across(c('prop', 'n'),
                                ~ dplyr::case_when(is.na(.x) ~ 0,
                                                   TRUE ~ .x))) %>%
    ggplot(aes(x = REF_label, y = prop, fill = mutation)) +
    geom_col() +
    ggh4x::facet_nested(age_bracket + id ~ mutation, scales = 'free_x') +
    theme_linedraw() +
    theme(axis.title.x = element_blank(),
          axis.text.x = ggtext::element_markdown(angle = 90, vjust = 0.5, family = 'cairo'),
          panel.grid.major.x = element_blank(),
          legend.position = 'none') +
    labs(y = 'relative contribution') +
    scale_fill_manual(values = mut_cols)
}

make_plots <- function(som_muts, som_mut_loads) {
  
  cat('\n\n**Density plot of somatic mutation VAFs**')
  plot_vafs_density(som_muts)
  
  cat('\n\n**Boxplot of somatic mutation VAFs**')
  plot_vafs_boxplot(som_muts)
  
  cat('\n\n**Barplot of somatic mutation load per ID, by level**\n')
  plot_ct_vs_mut_load_per_id_bar(som_mut_loads)
  
  cat('\n\n**Barplot of somatic mutation load per celltype, by level**\n')
  plot_ct_vs_mut_load_per_id_bar2(som_mut_loads)
  
  cat('\n\n**Heatmap of somatic mutation load per celltype, by level**\n')
  plot_celltype_vs_mut_load_per_id_heatmap(som_mut_loads)
  
  cat('\n\n**Heatmap of somatic mutation load per celltype vs n cells, by level**\n')
  print(plot_celltype_vs_mut_load_vs_n_cells_per_id_heatmap(som_mut_loads %>% dplyr::filter(lvl == 'donor_id')))
  
  cat('\n\n**Dotplot of somatic mutation load per Mb, compared between replicates of each donor**\n')
  print(plot_mut_load_btwn_replicates(som_mut_loads))
  
  cat('\n\n**Dotplot of n cells per level per celltype vs various metrics**\n')
  plot_n_cells_vs_metrics(som_mut_loads)
  
}
```

SComatic was run on scRNA-seq rheumatoid arthritis data using a modified version of kp9 and jc48’s [Nextflow implementation](https://github.com/Teichlab/scripts/tree/bd17ed22ec9e669056862d28c5f12a7e5033e68c/genotyping/scomatic/nextflow). 
Their implementation was intended to capture all germline and somatic variants in order to reconstruct tissue phylogeny, using both shared and private mutations for lineage tracing. In order to achieve this, they set `max_cell_types` to 1e6. I set it to 1. I also forced the pipeline to publish the celltype-split BAMs for each patient, which will be used to assess coverage of the driver genes in each cell type. The pipeline can be used with both GEX and ATAC modalities, hence the `*-GEX` suffix on the output directories.

Here is an example of the output files for donor BRI-401: 

```
out/scomatic/BRI_401-GEX
├── BRI_401.calling.step2.intersect.tsv
├── BRI_401.calling.step2.pass.tsv
├── BRI_401.calling.step2.tsv
├── BRI_401.coverage_cell_count.per_chromosome.report.tsv
├── BRI_401.coverage_cell_count.report.tsv
├── cell_callable_sites
│   ├── BRI_401.B_cell_plasma_cell.SitesPerCell.tsv
│   ├── BRI_401.Endothelial_cell.SitesPerCell.tsv
│   ├── BRI_401.Myeloid_cell.SitesPerCell.tsv
│   ├── BRI_401.NK.SitesPerCell.tsv
│   ├── BRI_401.Stromal_cell.SitesPerCell.tsv
│   └── BRI_401.T_cell.SitesPerCell.tsv
└── celltype_bams
    ├── BRI_401.B_cell_plasma_cell.bam
    ├── BRI_401.B_cell_plasma_cell.bam.bai
    ├── BRI_401.Endothelial_cell.bam
    ├── BRI_401.Endothelial_cell.bam.bai
    ├── BRI_401.Myeloid_cell.bam
    ├── BRI_401.Myeloid_cell.bam.bai
    ├── BRI_401.NK.bam
    ├── BRI_401.NK.bam.bai
    ├── BRI_401.Stromal_cell.bam
    ├── BRI_401.Stromal_cell.bam.bai
    ├── BRI_401.T_cell.bam
    └── BRI_401.T_cell.bam.bai
```

### Load metadata

```{r sm_and_ct_info, results = F, message = F, class.source = 'fold-hide'}
# sample metadata
sm <-
  '../scomatic_vs_monopogen/out/metadata.tsv' %>%
  readr::read_tsv() %>%
  dplyr::select(-c(irods_bam_file, median_n_counts, median_n_genes, n_cells)) %>%
  dplyr::distinct() %>%
  dplyr::mutate(age_bracket = factor(age_bracket, levels = c('young', 'old')),
                id = ifelse(lvl == 'sample_id', paste0(donor_id, '_', id), id))

# celltype info
ct_info <-
  '../scomatic_vs_monopogen/out/celltypes_w_counts.tsv' %>%
  readr::read_tsv() %>%
  dplyr::mutate(donor_id2 = donor_id) %>%
  tidyr::pivot_longer(
    cols = c('donor_id', 'sample_id'),
    names_to = 'lvl',
    values_to = 'id'
  ) %>%
  dplyr::rename(donor_id = donor_id2) %>%
  dplyr::mutate(id = ifelse(lvl == 'sample_id', paste0(donor_id, '_', id), id)) %>%
  dplyr::group_by(lvl, id, donor_id, annotation_level1, annotation_level1_alias) %>%
  dplyr::summarise(n_cells_per_id_per_celltype = sum(n))

# sce object + coldata
sce <- 
  readRDS('/nfs/team205/vk8/projects/somatic_mut_benchmark/SKM_cells_data2benchmark_all.Rds')
col_data <- 
  sce@colData %>%
  tibble::as_tibble(rownames = 'CB') %>%
  dplyr::rename(sample_id = SampleID, donor_id = DonorID) %>%
  dplyr::mutate(sample_id = paste0(sample_id, '_', donor_id))

# cellranger qc
cr_qc <-
  '../scomatic_vs_monopogen/data/cellranger_web_summaries/counts.tsv' %>%
  readr::read_tsv() 

# celltypes of interest
cts_oi <- c('MuSC', 'ArtEC',  'CapEC', 'VenEC', 'SMC', 'Pericyte',  'FB', 'Macrophage')
cts_pal <-
  ct_info %>%
  dplyr::group_by(annotation_level1) %>%
  dplyr::summarise(n = sum(n_cells_per_id_per_celltype)) %>%
  dplyr::arrange(-n) %>%
  dplyr::mutate(oi = ifelse(annotation_level1 %in% cts_oi, 'oi', 'noi')) %>%
  {split(.$annotation_level1, .$oi)}
cts_pal$oi <- rev(RColorBrewer::brewer.pal(length(cts_oi), 'Dark2')) %>% setNames(cts_pal$oi)
cts_pal$oi[1] <- '#0370b0'
cts_pal$noi <- rep('darkgrey', length(cts_pal$noi)) %>% setNames(cts_pal$noi)
cts_pal <- c(cts_pal$oi, cts_pal$noi)
```

#### Donor-level metadata

```{r donor_sm_df, echo = F}
sm %>%
  dplyr::filter(lvl == 'donor_id') %>%
  knitr::kable("html") %>% 
  kableExtra::kable_styling("striped")
```

#### Sample-level metadata

```{r sample_sm_df, echo = F}
sm %>%
  dplyr::filter(lvl == 'sample_id') %>%
  knitr::kable("html") %>% 
  kableExtra::kable_styling("striped") 
```

#### Cellranger QC 

```{r cr_qc, echo = F}
cr_qc %>%
  knitr::kable("html") %>% 
  kableExtra::kable_styling("striped", full_width = F) 
```

#### Celltype-level counts

```{r ct_df, echo = F}
ct_info %>%
  knitr::kable("html") %>% 
  kableExtra::kable_styling("striped", full_width = F) %>% 
  kableExtra::scroll_box(width = "100%", height = "300px")
```

```{r plot_cts_vs_ids, eval = F}
p_dat <- 
  ct_info %>%
  dplyr::left_join(sm) %>%
  dplyr::filter(lvl == 'donor_id') %>%
  dplyr::group_by(annotation_level1) %>%
  dplyr::mutate(
    n_cells_per_celltype = sum(n_cells_per_id_per_celltype)) %>%
  dplyr::group_by(id) %>%
  dplyr::mutate(
    n_cells_per_id = sum(n_cells_per_id_per_celltype)) %>%
  dplyr::ungroup() %>%
  dplyr::mutate(
    id = forcats::fct_reorder(id, n_cells_per_id),
    annotation_level1_lab = 
      ifelse(
        annotation_level1 %in% cts_oi,
        paste0('**', annotation_level1, '**'),
        annotation_level1) %>%
      forcats::fct_reorder(-n_cells_per_celltype),
    annotation_level1 = forcats::fct_reorder(annotation_level1, -n_cells_per_celltype),
    dplyr::across(
      .cols = dplyr::starts_with('n_'), 
      .fns = ~ prettyNum(.x, big.mark = ",", scientific = FALSE),
      .names = '{.col}_label'))

p_ids_x_cts <-
  p_dat %>%
  ggplot(aes(x = annotation_level1_lab, y = id, 
             colour = annotation_level1)) +
  geom_point(aes(size = n_cells_per_id_per_celltype, 
                 alpha = n_cells_per_id_per_celltype)) +
  scale_alpha(range = c(0.5, 1)) +
  scale_y_discrete(position = 'right')  +
  scale_x_discrete(position = 'bottom', guide = guide_axis(angle = -90)) +
  scale_size_continuous(range = c(1, 8)) +
  scale_colour_manual(guide = 'none', values = cts_pal) +
  theme_linedraw() +
  theme(legend.position = 'left', legend.title = element_blank(),
        axis.text.x = ggtext::element_markdown(angle = 90, vjust = 0.5),
        axis.title = element_blank(),
        strip.placement = 'outside') +
  facet_grid(age_bracket ~ ., scales = 'free_y', space = 'free_y', 
             switch = 'y') 

p_ids <-
  p_dat %>%
  dplyr::mutate(annotation_level1 = factor(annotation_level1, levels = rev(names(cts_pal)))) %>%
  ggplot(aes(x = id, y = n_cells_per_id_per_celltype,
             fill = annotation_level1)) +
  geom_col() +
  geom_text(aes(x = id, y = n_cells_per_id, label = n_cells_per_id_label),
            size = 3.5,
            hjust = -0.1, vjust = 0.5) +
  coord_flip() +
  scale_y_continuous(
    expand = c(0, 0), 
    limits = c(0, 1.4 * max(p_dat$n_cells_per_id))) +
  scale_fill_manual(values = cts_pal) +
  theme_void() +
  theme(legend.position = 'none', 
        strip.text.y = element_blank()) +
  facet_grid(age_bracket ~ ., scales = 'free_y', space = 'free_y') 

p_cts <-
  p_dat %>%
  dplyr::distinct(annotation_level1, 
                  n_cells_per_celltype, 
                  n_cells_per_celltype_label) %>%
  ggplot(aes(x = annotation_level1, 
             y = n_cells_per_celltype, 
             fill = annotation_level1)) +
  geom_col() +
  geom_text(aes(x = annotation_level1, y = n_cells_per_celltype, 
                label = n_cells_per_celltype_label),
            size = 5,
            hjust = 0, vjust = -0.1, angle = 45) +
  scale_y_continuous(
    expand = c(0, 0), 
    limits = c(0, 1.2 * max(p_dat$n_cells_per_celltype))) +
  scale_fill_manual(values = cts_pal) +
  theme_void() +
  theme(legend.position = 'none') 

# plots
cowplot::plot_grid(p_ids_x_cts, p_ids, ncol = 2, align = 'h',
                   rel_widths = c(1, .3))
p_cts
```

#### Quality metrics

```{r qc_plots, fig.height = 8}
p_dat <-
  sce@colData %>%
  tibble::as_tibble() %>%
  dplyr::left_join(sm %>% dplyr::mutate(DonorID = donor_id)) %>%
  dplyr::mutate(DonorID = forcats::fct_reorder(
                  paste0(DonorID, '\n(', X10X_version, ')'), age)) %>%
  dplyr::group_by(DonorID) %>%
  dplyr::mutate(n_cells = dplyr::n()) %>%
  dplyr::ungroup() %>%
  tidyr::pivot_longer(cols = c('age', 'n_counts', 'n_genes', 'n_cells')) %>%
  dplyr::mutate(
    name = gsub('_', ' ', name),
    name = ifelse(name %in% c('n cells', 'age'), name, 
                  paste0('log10(', name, ')')) %>%
      factor(levels = c('age', 'n cells', 'log10(n counts)', 'log10(n genes)')),
    value = ifelse(name %in% c('n cells', 'age'), value, log10(value))) 

p_dat %>%
  ggplot(aes(x = DonorID, y = value, fill = age_bracket)) +
  geom_violin(data = p_dat %>% dplyr::filter(!(name %in% c('age', 'n cells'))),
              aes(fill = age_bracket, colour = age_bracket)) +
  geom_jitter(data = p_dat %>% dplyr::filter(!(name %in% c('age', 'n cells'))),
              height = 0, alpha = 0.02, size = 0.005) +
  geom_col(data = p_dat %>% 
             dplyr::filter(name %in% c('age', 'n cells')) %>%
             dplyr::distinct(DonorID, name, value, age_bracket)) +
  facet_grid(name ~ age_bracket, scales = 'free', space = 'free_x') +
  theme_linedraw() +
  theme(axis.title.y = element_blank(),
        legend.position = 'none') +
  xlab('Donor ID\n(10X version)')
```

### Load called mutations

```{r load_som_muts, eval = F}
scom <- xfun::cache_rds({
  
  # initiate lists
  scom <- list(
    all_muts = list(),
    som_muts = list(),
    coverages = list(),
    sc_muts = list(),
    annovar = list()
  )
  
  # load somatic mutations and coverages for each ID/lvl
  c('donor_id', 'sample_id') %>%
    purrr::walk(function(lvl) {
      
      dir <- paste0('out/', lvl)
      
      # initiate sublists
      scom$all_muts[[lvl]]  <<- list()
      scom$som_muts[[lvl]]  <<- list()
      scom$coverages[[lvl]] <<- list()
      scom$sc_muts[[lvl]]   <<- list()
      
      list.files(dir, pattern = '-GEX$') %>%
        purrr::walk(function(subdir) {
          
          id <- gsub('-GEX', '', subdir)
          file_prefix <- paste0(dir, '/', subdir, '/', id)
          col_suffix <- ifelse(lvl == 'sample_id', 
                               sub('.*?_', '', id),
                               id) %>% {paste0('_', .)}

          message(paste(lvl, id))
          
          # get all called muts
          scom$all_muts[[lvl]][[id]] <<-
            paste0(file_prefix, '.calling.step2.pass.tsv') %>%
            readr::read_tsv(comment = '##') %>%
            # separate longer the Cell_types column and pivot longer the
            # calling columns and then separate wider the INFO 
            # -> produces one row per variant per celltype
            tidyr::pivot_longer(dplyr::ends_with(col_suffix), names_to = 'ct', values_to = 'call') %>%
            tidyr::separate_rows(Cell_types, VAF, sep = ',') %>%
            tidyr::separate(
              'call', 
              into = paste0('ct_', c('DP', 'NC', 'CC', 'BC', 'BQ', 'BCf', 'BCr')),
              sep = '\\|',  remove = F, convert = T) %>%
            readr::type_convert() %>%
            dplyr::filter(ct == Cell_types) %>%
            # get celltype names
            dplyr::mutate(
              annotation_level1_alias = gsub(col_suffix, '', Cell_types)) 
          
          # get called somatic mutations (exclusive to celltypes)
          scom$som_muts[[lvl]][[id]] <<-
            scom$all_muts[[lvl]][[id]] %>%
            # get vars shared by <= max_n_celltypes
            dplyr::group_by(`#CHROM`, Start, End, REF, ALT) %>%
            dplyr::filter(dplyr::n_distinct(annotation_level1_alias) == 1) %>%
            # get median VAF per celltype
            dplyr::group_by(annotation_level1_alias) %>%
            dplyr::mutate(median_VAF_per_id_per_celltype = median(VAF)) %>%
            dplyr::ungroup() 
          
          # get callable sites per celltype
          scom$coverages[[lvl]][[id]] <<-
            paste0(file_prefix, '.coverage_cell_count.per_chromosome.report.tsv') %>%
            readr::read_tsv() %>%
            # get celltype names
            dplyr::mutate(
              annotation_level1_alias = gsub(col_suffix, '', Cell_types))
          
          # get annovar gene annotations
          scom$annovar[[lvl]][[id]] <<-
            paste0(dir, '/', subdir, '/annovar/', 
                   id, '.variants.annovar.hg38_multianno.csv') %>%
            readr::read_csv() 
          
          # get somatic mutations at single cell resolution
          scom$sc_muts[[lvl]][[id]] <<-
            paste0(dir, '/', subdir, '-genotypes') %>%
            list.files(full.names = T) %>%
            purrr::map(function(file) {
              # message(file)
              sc_df <-
                readr::read_tsv(file, ) %>%
                dplyr::filter(Cell_type_expected == Cell_type_observed,
                              ALT_expected == Base_observed) %>%
                # get celltype names
                dplyr::mutate(
                  annotation_level1_alias = gsub(
                    paste0('_', id), '', Cell_type_observed))
              
              # get callable sites per cell
              ct <- unique(sc_df$Cell_type_observed)
              if (length(ct) != 1) {
                message(paste(length(ct), 'celltypes in the single cell genotypes file! Should be 1!'))
                message(paste('File', file, 'will be skipped!'))
                sc_df <- tibble::tibble()
              } else {
                sc_coverage <- 
                  paste0(dir, '/', subdir, '/cell_callable_sites/', 
                         id, '.', ct, '.SitesPerCell.tsv') %>%
                  readr::read_csv()
                # add single cell-level coverage
                sc_df <- dplyr::left_join(sc_df, sc_coverage, by = 'CB')
              }
              
              # return
              sc_df
                
            }) %>%
            dplyr::bind_rows()
          
        })
    })
  
  # add sample and celltype metadata
  scom <-
    scom %>% 
    purrr::map(function(df_ls_ls) {
      df_ls_ls %>%
        purrr::map(function(df_ls) {
          df_ls[sapply(df_ls, function(x) dim(x)[1]) > 0] %>%
            dplyr::bind_rows(.id = 'id')
        }) %>%
        dplyr::bind_rows(.id = 'lvl') %>%
        # # add celltype info
        dplyr::left_join(ct_info) %>%
        # # add ID info
        dplyr::left_join(sm) %>%
        # add grouping structure
        dplyr::group_by(# sample info
                        lvl, id, donor_id, sex, age_bracket, age, X10X_version, n_cells_per_id,
                        # celltype info
                        annotation_level1_alias, annotation_level1, n_cells_per_id_per_celltype)
    })
  
  # return
  scom
      
}, file = 'scom.rds', rerun = do_rerun)

# assign dfs globally
all_muts <- xfun::cache_rds({scom$all_muts}, file = 'all_muts.rds', rerun = do_rerun)
som_muts <- xfun::cache_rds({scom$som_muts}, file = 'som_muts.rds', rerun = do_rerun)
coverages <- xfun::cache_rds({scom$coverages}, file = 'coverages.rds', rerun = do_rerun)
sc_muts <- xfun::cache_rds({scom$sc_muts}, file = 'sc_muts.rds', rerun = do_rerun)
annovar <- xfun::cache_rds({scom$annovar}, file = 'annovar.rds', rerun = do_rerun)
# list2env(scom, globalenv())
```

```{r load_up}
all_muts <- readRDS('reports/mutation_load_cache/all_muts.rds')
som_muts <- readRDS('reports/mutation_load_cache/som_muts.rds')
coverages <- readRDS('reports/mutation_load_cache/coverages.rds')
sc_muts <- readRDS('reports/mutation_load_cache/sc_muts.rds')
annovar <- readRDS('reports/mutation_load_cache/annovar.rds')
```

`*.calling.step2.pass.tsv` represents the somatic mutation calls at celltype resolution. These files look like this:

```{r pass_df, echo = F}
head(som_muts) %>%
  knitr::kable("html") %>% 
  kableExtra::kable_styling("striped") %>% 
  kableExtra::scroll_box(width = "100%", height = "300px")
```


The columns have the following definitions:

```{r pass_df_cols, echo = F}
'out/donor_id/339C-GEX/339C.calling.step2.pass.tsv' %>% 
  readr::read_delim(delim = ',Description=', skip = 1, col_names = F) %>%
  dplyr::filter(grepl('##INFO', X1)) %>%
  dplyr::transmute(
    col = gsub('##INFO=', '', X1),
    definition = gsub('>', '', X2)
  ) %>%
  knitr::kable("html") %>% 
  kableExtra::kable_styling("striped")
```


`*.coverage_cell_count.report.tsv` files look like this:

```{r cov_df, echo = F}
head(coverages, 10) %>%
  knitr::kable("html") %>% 
  kableExtra::kable_styling("striped")
```


The columns have the following definitions:

```{r cov_df_cols, echo = F}
tibble::tribble(
  ~col, ~definition,
  'Cell_types', 'celltype (Z)',
  'Cov', 'coverage (n)',
  'DP', 'number of sites in celltype Z with at least n reads',
  'NC', 'number of sites in celltype Z in at least n cells'
) %>%
  knitr::kable("html") %>% 
  kableExtra::kable_styling("striped")
```


`*.single_cell_genotypes.tsv` files look like this:

```{r sc_df, echo = F}
head(sc_muts) %>%
  knitr::kable("html") %>% 
  kableExtra::kable_styling("striped") %>% 
  kableExtra::scroll_box(width = "100%", height = "300px")
```

### Calculate somatic mutation loads within celltypes

Somatic mutation load was calculated according to [issue #23](https://github.com/cortes-ciriano-lab/SComatic/issues/23) on the SComatic GitHub.

> Regarding the "callable sites" question, it is important to take into account that there are two types of values when we speak about coverage (Cov column):
>
> 1. one based on the number of cells with at least one read at a given position (NC column)
>
> 2. another based on the depth or number of reads at each site (DP column).
>
> To clarify this concept and the file format, it is much easier to understand it with one example:
>
> In the B_memory showed in your attached figure, when Cov == 5:
>
> * DP shows the number of sites with 5 reads = 472874
>
> * NC shows the number of sites with 5 cells = 2242427
>
> You have this site counting value for each coverage (up to 150 by default). By looking at these values, you can get the number of callable sites based on the minimum coverage (Cov) that you want. For instance, if you want to get the callable sites with at least 10 cells, you should sum the column NC for all rows in the cell type with Cov >= 10.
>
> In our manuscript, we computed the mutation load at cell-type resolution by using a minimum Cov >= 5 and the next formula:
>
> `(# somatic mutations in the cell type Z) / (# callable sites in the cell type Z)`

`# somatic mutations in the cell type Z` can be calculated by "counting the number of rows in the file sample.calling.step2.pass.tsv for each cell type without any more filtering".

So, for celltype *Z* of ID *X*:

*somatic mutation load in Z = (n somatic mutations in Z) / (n callable sites in Z)*

* n somatic mutations in Z = `X.calling.step2.pass.tsv %>% count(Cell_types)`
* n callable sites in celltype Z = `X.coverage_cell_count.report.tsv %>% filter(NC > min_n_cells_per_site) %>% sum(NC per Z)`

**Steps:**

1. Get n somatic mutations per celltype per ID. Somatic mutations must be filtered to exclusive mutations within each celltype. Because SComatic was run with `max_cell_types = 100000`, the output will also contain germline and developmental (mosaic) mutations, shared between celltypes. We are interested in celltype-specific somatic mutations, so must remove these. This can be done by filtering the `Cell_types` column to only variants reported in one celltype. The instances of each celltype in the `Cell_types` column can then be counted to get n somatic mutations.
2. Get n callable sites per celltype per ID by summing the `NC` column of the coverage report.
3. Calculate mutational load per celltype per ID as `n_som_muts` / `n_callable_sites`. 

This will be done at different cut-offs for coverage depth, number of cells, and label granularity.

```{r, echo = F}
tibble::tribble(
  ~arg, ~definition, ~default,
  'max_n_celltypes', 'Maximum n celltypes sharing the variant for it to be called as a somatic mutation.', '1',
  'min_n_cells', 'Minimum n cells in which a variant is detected for it to be called as a somatic mutation.', '5',
  'min_coverage_depth', 'Minimum n reads at a site at which a variant is detected for it to be called as a somatic mutation', '5',
  'min_n_cells_per_celltype_across_ids', 'Minimum number of cells for a celltype in all IDs for it to be interrogated for somatic mutations. If any ID has less than this number of cells in a celltype, the celltype is not interrogated. This is to ensure that only well-represented celltypes are included in comparisons.', '1',
  'min_n_callable_sites_per_celltype', 'Minimum number of callable sites for a cell type in an ID for it to be included.', '500,000',
  'label', 'The labelling level at which to group cells into celltypes. annotation_level0 is coarsest, annotation_level2 is finest.', 'annotation_level1'
) %>%
  knitr::kable("html") %>% 
  kableExtra::kable_styling("striped") 
```

```{r get_som_mut_loads}
get_som_mut_loads <- 
  function(
    som_muts,
    coverages,
    ct_info,
    min_n_cells = 5,
    min_coverage_depth = 5,
    min_n_cells_per_celltype_across_ids = 1,
    min_n_callable_sites_per_celltype = 500000,
    remove_XY = T,
    label = 'annotation_level1' # annotation_level2 = finer, annotation_level0 = coarser
  ) {
    # min_n_cells=5; min_coverage_depth=5;min_n_cells_per_celltype_across_ids=1;min_n_callable_sites_per_celltype=500000;remove_XY=T
    
    # check that no celltype names contain commas
    if(any(grepl(',', ct_info$annotation_level1))) stop("Commas in the celltype names!")
    
    # 1. get n somatic mutations
    som_mut <-
      som_muts %>%
      # i. filter out variants...
      #    - in cell types with insufficient coverage 
      #    - in cell types insufficient cells
      dplyr::filter(ct_DP >= min_coverage_depth,
                    ct_NC >= min_n_cells) %>%
      # ii. optionally filter out variants on the sex chromosomes
      { if (remove_XY == T) dplyr::filter(., !(`#CHROM` %in% c('X', 'Y'))) else . } %>%
      # ii. count mutations per celltype
      dplyr::count(name = 'n_som_muts') 
    
    # 2. get n callable sites per celltype
    coverage <-
      coverages %>%
      # i. sum n callable sites per celltype 
      #    (with coverage >= min_n_cells, on autosomes)
      dplyr::filter(Cov >= min_n_cells) %>%
      { if(remove_XY == T) dplyr::filter(.,  !(CHROM %in% c('X', 'Y'))) else . } %>%
      dplyr::summarise(n_callable_sites = sum(NC)) %>%
      # ii. filter n callable sites >= min_n_callable_sites_per_celltype
      dplyr::filter(n_callable_sites >= min_n_callable_sites_per_celltype) %>%
      dplyr::group_by(lvl, id)
    
    # 3. calculate somatic mutation load
    som_mut_loads <-
      # i. join n somatic mutations and coverage by celltype
      dplyr::inner_join(som_mut, coverage) %>%
      # ii. calculate som mut load and som mut rate:
      #       if no callable sites ~ NA
      #       if callable sites ~ n_som_muts / n_callable_sites
      dplyr::mutate(
        som_mut_load = ifelse(
          n_callable_sites == 0, NA,
          n_som_muts / n_callable_sites),
        som_mut_load_per_Mb = som_mut_load * 1e6,
        som_mut_load_per_genome = som_mut_load * 3.2e9,
        som_mut_load_per_genome_per_yr = som_mut_load_per_genome / age) %>%
      # iii. filter out celltypes with insufficient cells in any one ID
      dplyr::inner_join(
        get_hq_cts(ct_info, min_n_cells_per_celltype_across_ids))
    
    # return
    som_mut_loads
    
    }
```

The function is applied to the somatic mutations at different cutoff stringencies. 

```{r}
som_mut_loads <-
  get_som_mut_loads(
    som_muts, coverages, ct_info,
    min_n_cells = 5,
    min_coverage_depth = 5,
    min_n_cells_per_celltype_across_ids = 50,
    min_n_callable_sites_per_celltype = 500000
  )

som_mut_loads %>% 
    dplyr::filter(lvl == 'sample_id', annotation_level1 %in% cts_oi) %>% 
    ggplot(aes(y = som_mut_load_per_Mb, x = age_bracket, fill = age_bracket)) + 
    geom_boxplot() + 
    geom_point(alpha = 0.5, position = 'dodge') + 
    facet_grid(~ annotation_level1)

# save 
som_mut_loads %>%
  readr::write_tsv('out/som_mut_loads.tsv')
```

```{r eval = F, echo = F}
df <- 
  som_mut_loads %>%
  dplyr::arrange(-som_mut_load_per_Mb)  %>%
  dplyr::ungroup() %>%
  dplyr::select(
    lvl, id, annotation_level1_alias,
    n_callable_sites, n_som_muts, som_mut_load_per_Mb, n_cells_per_id_per_celltype
  )
df %>%
  kableExtra::kbl() %>%
  kableExtra::kable_styling("striped") %>% 
  colour_kbl_column(df, 'n_callable_sites') %>%
  colour_kbl_column(df, 'n_som_muts') %>%
  colour_kbl_column(df, 'som_mut_load_per_Mb') %>%
  colour_kbl_column(df, 'n_cells_per_id_per_celltype') %>% 
  kableExtra::scroll_box(width = "100%", height = "400px")
```

### Plots

```{r plots_final, results = 'asis', echo = F}
make_plots(som_muts, som_mut_loads)
```

### Plots of celltypes of interest

We are primarily interested in 8 celltypes:

* `r paste(cts_oi, collapse = '\n\n* ')`

```{r plots_cts_oi, results = 'asis', echo = F}
make_plots(
  som_muts %>% dplyr::filter(lvl == 'donor_id', annotation_level1_alias %in% cts_oi), 
  som_mut_loads %>% dplyr::filter(lvl == 'donor_id', annotation_level1_alias %in% cts_oi))
```

### Mutational signatures

```{r plot_mutational_signatures, echo = F, fig.width = 10, fig.height = 10, eval = F}
som_muts %>%
  dplyr::filter(lvl == 'donor_id') %>%
  plot_mutational_signatures()
```

### Mutation load at single cell resolution

How do we calculate the mutation load of every individual cell?

This question was asked and answered in [issue #30](https://github.com/cortes-ciriano-lab/SComatic/issues/30) on the SComatic GitHub:

> The basic strategy would be:
>
> 1. Compute how many rows per cell accomplish the `ALT_expected == Base_observed` and `Cell_type_expected == Cell_type_observed`. Basically, the number of mutations per cell.
> 2. It is essential to consider the number of callable sites per cell, as it will affect the number of mutations detected. To perform this correction, you will need to compute the number of callable sites per cell using `scripts/GetCallableSites/GetAllCallableSites.py`. You can use these callable sites to compute for example the mutation load per cell and MB (# Mutations per cell / # Callable sites per cell).
> 3. Plot the resulting values. I would ignore those cells with a very low number of callable sites. Generally, you will see an enrichment of cells at 0. This is due to the low number of callable sites per cell in this type of approaches (scRNA-seq) and a low mutation load (depending on the cancer or cell type).

```{r sc_som_mut_loads}
sc_som_mut_loads <- xfun::cache_rds({
  sc_muts %>%
    # only sites in the callable sites at cell type level
    dplyr::inner_join(
      all_muts %>%
        dplyr::ungroup() %>%
        dplyr::select(lvl, id, `#CHROM`, Start, End, REF, ALT, annotation_level1_alias)
    ) %>%
    dplyr::group_by(CB, SitesPerCell, .add = T) %>%
    dplyr::count(name = 'n_som_muts') %>%
    dplyr::mutate(som_mut_load = n_som_muts / SitesPerCell,
                  som_mut_load_per_Mb = som_mut_load * 1e6) 
}, file = 'sc_som_mut_loads.rds')
```

```{r}
GeomSplitViolin <- 
  ggproto("GeomSplitViolin", GeomViolin, draw_group = function(self, data, ..., draw_quantiles = NULL) {
  data <- transform(data, xminv = x - violinwidth * (x - xmin), xmaxv = x + violinwidth * (xmax - x))
  grp <- data[1, "group"]
  newdata <- plyr::arrange(transform(data, x = if (grp %% 2 == 1) xminv else xmaxv), if (grp %% 2 == 1) y else -y)
  newdata <- rbind(newdata[1, ], newdata, newdata[nrow(newdata), ], newdata[1, ])
  newdata[c(1, nrow(newdata) - 1, nrow(newdata)), "x"] <- round(newdata[1, "x"])

  if (length(draw_quantiles) > 0 & !scales::zero_range(range(data$y))) {
    stopifnot(all(draw_quantiles >= 0), all(draw_quantiles <=
      1))
    quantiles <- ggplot2:::create_quantile_segment_frame(data, draw_quantiles)
    aesthetics <- data[rep(1, nrow(quantiles)), setdiff(names(data), c("x", "y")), drop = FALSE]
    aesthetics$alpha <- rep(1, nrow(quantiles))
    both <- cbind(quantiles, aesthetics)
    quantile_grob <- GeomPath$draw_panel(both, ...)
    ggplot2:::ggname("geom_split_violin", grid::grobTree(GeomPolygon$draw_panel(newdata, ...), quantile_grob))
  }
  else {
    ggplot2:::ggname("geom_split_violin", GeomPolygon$draw_panel(newdata, ...))
  }
})
geom_split_violin <- function(mapping = NULL, data = NULL, stat = "ydensity", position = "identity", ..., 
                              draw_quantiles = NULL, trim = TRUE, scale = "area", na.rm = FALSE, 
                              show.legend = NA, inherit.aes = TRUE) {
  layer(data = data, mapping = mapping, stat = stat, geom = GeomSplitViolin, 
        position = position, show.legend = show.legend, inherit.aes = inherit.aes, 
        params = list(trim = trim, scale = scale, draw_quantiles = draw_quantiles, na.rm = na.rm, ...))
}
sc_som_mut_loads %>%
  dplyr::left_join(sm %>% dplyr::select(id, X10X_version)) %>%
  dplyr::filter(annotation_level1 %in% cts_oi) %>%
  ggplot(aes(x = annotation_level1, 
             y = som_mut_load_per_Mb, 
             fill = age_bracket)) + 
  geom_split_violin() +
  scale_x_discrete(guide = guide_axis(angle = -90)) +
  theme_bw() +
  facet_wrap(~ X10X_version)
```

We can look at the extreme cells to see how their somatic mutation loads are being derived.

```{r echo = F}
df <- 
  sc_som_mut_loads %>%
  dplyr::arrange(-som_mut_load_per_Mb)  %>%
  head(100) %>%
  dplyr::ungroup() %>%
  dplyr::select(
    lvl, id, annotation_level1_alias,
    n_som_muts, SitesPerCell, som_mut_load_per_Mb, n_cells_per_id_per_celltype
  )
df %>%
  kableExtra::kbl() %>%
  kableExtra::kable_styling("striped") %>% 
  colour_kbl_column(df, 'SitesPerCell') %>%
  colour_kbl_column(df, 'n_som_muts') %>%
  colour_kbl_column(df, 'som_mut_load_per_Mb') %>%
  colour_kbl_column(df, 'n_cells_per_id_per_celltype') %>% 
  kableExtra::scroll_box(width = "100%", height = "400px")
```

```{r plot_sc_som_mut_loads, echo = F, fig.width = 10}
sc_som_mut_loads %>% 
  dplyr::group_by(lvl, id, annotation_level1) %>% 
  dplyr::mutate(median_som_mut_load_per_Mb = median(som_mut_load_per_Mb)) %>%
  dplyr::filter(lvl == 'donor_id', annotation_level1_alias %in% cts_oi) %>%
  ggplot(aes(x = "", y = som_mut_load_per_Mb)) +
  geom_violin() +
  geom_jitter(alpha = 0.2, size = 0.2, height = 0,
              aes(colour = SitesPerCell)) +
  geom_hline(aes(yintercept = median_som_mut_load_per_Mb)) +
  theme(axis.text.x = element_text(angle = 270, hjust = 0, vjust = -0.5)) +
  scale_colour_viridis_b() +
  coord_flip() +
  ggh4x::facet_nested(age_bracket + id ~ annotation_level1) 
```

```{r plot_sc_som_mut_load_vs_variables}
sc_som_mut_loads %>%
  ggplot(aes(x = SitesPerCell, y = som_mut_load_per_Mb, 
             colour = donor_id)) +
  geom_point() +
  facet_grid(~ lvl)
```

```{r}
cb_x_som_mut <-
  tibble::tibble(CB = colnames(sce)) %>%
  dplyr::left_join(
    sc_som_mut_loads %>%
      dplyr::filter(lvl == 'donor_id') %>%
      dplyr::ungroup() %>%
      dplyr::transmute(
        CB = CB %>% stringi::stri_replace_last_fixed('_', '-'), 
        som_mut_load_per_Mb)) %>%
  tibble::column_to_rownames('CB')
sce$som_mut_load_per_Mb <- 
  cb_x_som_mut$som_mut_load_per_Mb
scater::plotReducedDim(sce, 'UMAP', colour_by = 'som_mut_load_per_Mb')
```

### Building a GLM

```{r glm}
p1 <-
  som_mut_loads %>%
  dplyr::left_join(sm) %>% 
  dplyr::filter(lvl == 'sample_id', annotation_level1 %in% cts_oi) %>% 
  ggplot(aes(x = n_cells_per_id_per_celltype, y = n_som_muts, colour = X10X_version, shape = age_bracket)) + 
  geom_point()
p2 <-
  som_mut_loads %>%
  dplyr::left_join(sm) %>% 
  dplyr::filter(lvl == 'sample_id', annotation_level1 %in% cts_oi) %>% 
  ggplot(aes(x = n_callable_sites, y = n_som_muts, colour = X10X_version, shape = age_bracket)) + 
  geom_point()
p3 <-
  som_mut_loads %>%
  dplyr::left_join(sm) %>% 
  dplyr::filter(lvl == 'sample_id', annotation_level1 %in% cts_oi) %>% 
  ggplot(aes(x = n_callable_sites, y = n_som_muts, colour = X10X_version, shape = age_bracket)) + 
  geom_point()
som_mut_loads %>%
  dplyr::left_join(sm) %>% 
  dplyr::inner_join(
    sce@colData %>% 
      tibble::as_tibble() %>% 
      dplyr::transmute(annotation_level1, id = paste0(DonorID, '_', SampleID), n_counts) %>%
      dplyr::group_by(id, annotation_level1) %>%
      dplyr::summarise(avg_n_counts = mean(n_counts))) %>%
  dplyr::filter(lvl == 'sample_id', annotation_level1 %in% cts_oi) %>% 
  ggplot(aes(x = avg_n_counts, y = som_mut_load_per_Mb, colour = X10X_version, shape = age_bracket)) + 
  geom_point()
som_mut_loads %>%
    dplyr::left_join(sm) %>% 
    dplyr::inner_join(
        sce@colData %>% 
            tibble::as_tibble() %>% 
            dplyr::transmute(annotation_level1, id = paste0(DonorID, '_', SampleID), n_counts) %>%
            dplyr::group_by(id, annotation_level1) %>%
            dplyr::summarise(avg_n_counts = mean(n_counts))) %>%
    dplyr::filter(lvl == 'sample_id', annotation_level1 %in% cts_oi, X10X_version == "3'v2") %>% 
    ggplot(aes(x = avg_n_counts, y = som_mut_load_per_Mb, colour = X10X_version, shape = age_bracket)) + 
    geom_point()

# n_callable_sites > 5e6
som_mut_loads %>%
    dplyr::left_join(sm) %>% 
    dplyr::filter(lvl == 'sample_id', annotation_level1 %in% cts_oi, X10X_version == "3'v2", n_callable_sites > 5e6) %>% 
    ggplot(aes(x = n_callable_sites, y = som_mut_load_per_Mb, colour = age_bracket, shape = age_bracket)) + 
    geom_point() 

glm_dat <-
  som_mut_loads %>%
  dplyr::filter(annotation_level1 %in% cts_oi) %>%
  dplyr::left_join(sm) %>%
  dplyr::inner_join(
        sce@colData %>% 
            tibble::as_tibble() %>% 
            dplyr::transmute(annotation_level1, id = paste0(DonorID, '_', SampleID), n_counts) %>%
            dplyr::group_by(id, annotation_level1) %>%
            dplyr::summarise(avg_n_counts = mean(n_counts))) %>%
  dplyr::transmute(
    n_som_muts,
    som_mut_load_per_Mb,
    age_bracket = factor(age_bracket, levels = c('young', 'old')),
    n_cells_per_id_per_celltype,
    n_callable_sites,
    avg_n_counts,
    X10X_version)
glm_obj <-
  glm(
  n_som_muts ~
    # donor-level effects
    #age_bracket + annotation_level1 +
    age_bracket * annotation_level1 + 
    annotation_level1 + 
    avg_n_counts,
   # n_cells_per_id_per_celltype + 
    # cell type-level effects
    # n_callable_sites +
    # technical effects
    #X10X_version, 
  family = 'poisson',
  data = glm_dat
)
summary(glm_obj)
```

```{r fig.height = 8}
annovar %>%
  # get somatic muts only
  dplyr::inner_join(
    som_muts %>% dplyr::select(Chr = `#CHROM`, Start, End, Ref = REF, Alt = ALT)) %>%
  # get celltypes of interest
  dplyr::filter(annotation_level1 %in% cts_oi) %>%
  # get counts as fraction of all muts
  dplyr::mutate(n = dplyr::n(), `genomic region` = Func.refGene) %>%
  # get counts overall
  dplyr::group_by(`genomic region`) %>%
  dplyr::mutate(n = dplyr::n()) %>%
  dplyr::ungroup() %>%
  dplyr::mutate(`genomic region` = forcats::fct_reorder(`genomic region`, -n)) %>%
  ggplot(aes(x = `genomic region`, fill = age_bracket)) +
  geom_bar() +
  theme_linedraw() +
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0.5),
        legend.position = 'none',
        panel.grid = element_blank()) +
  facet_grid(annotation_level1 ~ age_bracket)
```

```{r, fig.height = 4}
# deleterious mutations
annovar %>%
  # get somatic muts only
  dplyr::inner_join(
    som_muts %>% 
      dplyr::select(Chr = `#CHROM`, Start, End, Ref = REF, Alt = ALT)) %>% 
  # classify deleterious vs non deleterious
  dplyr::mutate(
    `deleterious?` = 
      ifelse((is.na(MetaLR_pred) & is.na(MetaSVM_pred)), 'unknown',
             ifelse((MetaLR_pred == "T" | MetaSVM_pred == 'T'), 'deleterious',
             'not deleterious'))) %>%
  dplyr::group_by(`deleterious?`, .add = T) %>%
  ggplot(aes(x = donor_id, fill = `deleterious?`)) + 
  geom_bar() +
  theme_linedraw() +
  theme(panel.grid = element_blank()) +
  scale_fill_manual(values = c('deleterious' = '#c2625b',
                               'not deleterious' = '#72b86e',
                               'unknown' = 'darkgrey')) +
  facet_grid(~ age_bracket, scales = 'free_x', space = 'free_x')
```

### Final plots

```{r boxplot, echo = F, fig.width = 7, fig.height = 3}
# boxplot som_mut_load_per_Mb vs age
som_mut_loads %>%
  dplyr::filter(lvl == 'sample_id') %>%
  ggplot(aes(x = age, y = som_mut_load_per_Mb, group = donor_id)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_jitter(height = 0, size = 0.1, aes(colour = age_bracket)) +
  theme_linedraw() +
  theme(panel.grid.major.x  = element_blank(),
        panel.grid.minor.x = element_blank())
```

```{r sc_and_ct_levels, fig.height = 8}
# metrics at cell type and single cell level
p_dat <-
  som_mut_loads %>%
  tidyr::pivot_longer(
    cols = c('n_callable_sites', 
             'som_mut_load_per_Mb', 
             'n_som_muts')) %>%
  dplyr::mutate(level = 'cell type-level') %>%
  dplyr::bind_rows(
    sc_som_mut_loads %>% 
      dplyr::transmute(som_mut_load_per_Mb,
                       n_callable_sites = SitesPerCell,
                       n_som_muts = n_som_muts,
                       level = 'single cell-level') %>%
      tidyr::pivot_longer(
        cols = c('som_mut_load_per_Mb', 
                 'n_callable_sites',
                 'n_som_muts'))) %>%
  dplyr::filter(lvl == 'sample_id', annotation_level1 %in% cts_oi) %>%
  dplyr::group_by(annotation_level1) %>%
  dplyr::mutate(name = factor(gsub('_', ' ', name), 
                              levels = c('n callable sites', 'n som muts', 'som mut load per Mb')))

p_dat %>%
  ggplot(aes(x = age_bracket, y = value)) +
  geom_boxplot(data = p_dat %>% dplyr::filter(level == 'cell type-level'),
               outlier.alpha = 0) + 
  geom_jitter(data = p_dat %>% dplyr::filter(level == 'cell type-level'),
              aes(colour = age_bracket), height = 0, width = 0.3) + 
  geom_violin(data = p_dat %>% dplyr::filter(level == 'single cell-level'),
              draw_quantiles = 0.5) +
  geom_jitter(data = p_dat %>% dplyr::filter(level == 'single cell-level'),
              aes(colour = age_bracket), height = 0, width = 0.3, size = 0.1) + 
  theme_linedraw() +
  ggh4x::facet_nested(level + name ~ annotation_level1, scales = 'free_y') +
  labs(y = '', x = 'age bracket') + 
  theme(legend.position = 'none')
```

```{r metrics_vs_corrections}
# somatic mutation load vs n cells in the cell type
# somatic mutation load vs average n counts
# somatic mutation load vs n callable sites
p_ls <-
  som_mut_loads %>%
  # add qc metrics from sce
  dplyr::left_join(
    col_data %>%
      dplyr::group_by(sample_id, annotation_level1) %>%
      dplyr::summarise(dplyr::across(.cols = dplyr::starts_with('n_'),
                                     .fns = mean, .names = 'avg_{.col}'))
  ) %>%
  dplyr::rename(n_cells = n_cells_per_id_per_celltype) %>%
  dplyr::filter(lvl == 'sample_id') %>% 
  tidyr::pivot_longer(cols = c('n_callable_sites', 
                               'n_cells', 
                               'avg_n_counts'), 
                      names_to = 'x_var', values_to = 'x_value') %>%
  tidyr::pivot_longer(cols = c('n_som_muts', 'som_mut_load_per_Mb'),
                      names_to = 'y_var', values_to = 'y_value') %>%
  dplyr::arrange(y_var, x_var) %>%
  dplyr::mutate(split = paste0(y_var, '_vs_', x_var)) %>%
  dplyr::group_by(split) %>%
  split(., .$split) %>%
  purrr::map(function(df) {
    p <-
      df %>%
      ggplot(aes(x = x_value, y = y_value,
                 colour = donor_id, shape = age_bracket)) +
      geom_point() +
      theme_bw() +
      theme(aspect.ratio = 1) +
      labs(x = gsub('_', ' ', unique(df$x_var)), 
           y = gsub('_', ' ', unique(df$y_var)))
    p_legend <<- cowplot::get_legend(p)
    p + theme(legend.position = 'none')
  }) 
p_ls %>%
  patchwork::wrap_plots() 
patchwork::wrap_elements(p_legend)
```

```{r final_vafs, fig.height = 7}
som_muts %>%
  dplyr::filter(annotation_level1 %in% cts_oi, lvl == 'donor_id') %>%
  dplyr::group_by(id, annotation_level1) %>%
  dplyr::mutate(n = dplyr::n()) %>%
  ggplot(aes(x = VAF, label = n, fill = age_bracket)) +
  geom_density() +
  geom_vline(aes(xintercept = median_VAF_per_id_per_celltype), colour = 'grey') +
  geom_text(aes(x = 0.6, y = Inf), vjust = 1.3) +
  ggh4x::facet_nested(age_bracket + id ~ annotation_level1, scales = 'free_y') +
  theme_linedraw() +
  theme(axis.text.x = element_text(angle = -90, hjust = 0),
        axis.text.y = element_blank(),
        strip.text.y.right = element_text(angle = 0, hjust = 0),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        legend.position = 'none')
```

```{r final_vafs_superimposed, fig.height = 7}
som_muts %>%
  dplyr::filter(annotation_level1 %in% cts_oi, lvl == 'donor_id') %>%
  dplyr::group_by(annotation_level1) %>%
  dplyr::mutate(n = dplyr::n(),
                median_VAF_per_celltype = median(VAF)) %>%
  ggplot(aes(x = VAF, label = n, fill = age_bracket, group = id)) +
  geom_density(alpha = 0.2) +
  geom_vline(aes(xintercept = median_VAF_per_celltype)) +
  geom_text(aes(x = 0.6, y = Inf), vjust = 1.3) +
  ggh4x::facet_grid2(annotation_level1 ~ age_bracket, 
                     scales = 'free_y', independent = "y",
                strip = ggh4x::strip_vanilla(size = "variable")) +
  theme_linedraw() +
  theme(axis.text.x = element_text(angle = -90, vjust = 0.5),
        axis.title.y = element_blank(),
        axis.text.y = element_text(size = 5),
        axis.ticks.y = element_blank(),
        strip.text.y.right = element_text(angle = 0, hjust = 0),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        legend.position = 'none')
```
