---
title: "Mutation loads in rheumatoid arthritis data from SComatic"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: true
    code_folding: show
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
params:
    rerun: true
---


```{r setup, include = F, message = F, warning = F, class.source = 'fold-hide'}
# wd <- ifelse(Sys.info()['nodename'] == "mib119104s", '~/Volumes/', '') %>% paste0('/lustre/scratch125/casm/team268im/at31/RA_som_mut/scomatic/') ; setwd(wd)
# rmarkdown::render('reports/driver_coverage.Rmd', output_file = 'driver_coverage.html', output_dir = 'reports/', params = list(rerun = T))

knitr::opts_knit$set(root.dir = '../')
knitr::opts_chunk$set(
  warning = FALSE, 
  dpi = 300, 
  message = FALSE,
  cache.path = 'reports/driver_coverage_cache/') 

# libraries
library(magrittr)
library(ggplot2)
library(gridExtra)

# dev.off() if there is a device
dev_off_if <- function() {if (length(dev.list()) != 0) { dev.off() }}
```

### Drivers

Genes of interest ('drivers') in rheumatoid arthritis are loaded. Their coverage in the dataset, across samples and cell types, will be checked.

```{r drivers, warning=F, message=F}
drivers <-
  readr::read_tsv('data/driver_genes/driver_gene_coords_for_coverage.tsv')
knitr::kable(drivers)
```

### Calculate coverage, by sample and by sample-x-celltype

We use `samtools depth` to get the coverage of each site in the genes of interest, both across each sample and within each cell type in each sample.

```{bash cov, eval=F}
# activate conda env
mamba activate jupy

# dirs
input_dir=/lustre/scratch126/casm/team268im/mp34/scRNAseq_data/RA_ZhangEtal2023/cellranger_output
ct_input_dir=out/Zhang2023/default_thresholds/
output_dir=out/Zhang2023/coverage/
drivers=data/driver_genes/driver_gene_coords_for_coverage.tsv
mkdir -p $output_dir

# iterate over each bam file and pileup specific regions
(
  echo -e 'chr\tpos\tdepth\tgene\tsample' ;
  for bam in $(find $input_dir -iname 'possorted_genome_bam.bam'); do
  	sample=$(echo $bam | cut -d "/" -f 10 | sed 's/\-/_/g')
    while read -r gene coords ; do
      samtools depth -a -r $coords -Q 30 $bam |
      awk -v sample=$sample -v gene=$gene '{print $0"\t"gene"\t"sample}'
    done < <(sed 1d $drivers) ;
  done
) | cat > $output_dir/drivers_samtools_depth.tsv

# iterate over celltype-specific bam files and pileup specific regions
(
  echo -e 'chr\tpos\tdepth\tgene\tsample\tcelltype' ;
  for sample_dir in $ct_input_dir/*-GEX ; do
    sample=$(basename $sample_dir | sed 's/\-GEX$//g')
    for ct_bam in $sample_dir/celltype_bams/*.bam ; do
      ct=$(basename $ct_bam | cut -f2 -d.)
      while read -r gene coords ; do
        samtools depth -a -r $coords -Q 30 $ct_bam |
        awk -v sample=$sample -v gene=$gene -v ct=$ct \
        '{print $0"\t"gene"\t"sample"\t"ct}'
      done < <(sed 1d $drivers) ;
    done
  done
) | cat > $output_dir/drivers_samtools_depth_by_celltype.tsv
```

We read in the results at both levels.

```{r read_cov}
cov <- xfun::cache_rds({
  cov_files <-
    list.files(
      'out/Zhang2023/coverage/', 
      pattern = 'drivers_samtools_depth.*.tsv',
      full.names = T) 
  cov <-
    cov_files %>%
    setNames(cov_files %>% basename() %>% tools::file_path_sans_ext()) %>%
    purrr::map(readr::read_tsv) %>%
    dplyr::bind_rows(.id = 'lvl') %>%
    dplyr::mutate(chr = as.character(chr))
  cov
}, file = 'cov.rds', rerun = params$rerun)
```

### Annotate exons and introns

We use the reference object from DNDSCV to delimit intronic and exonic regions of the genes.

```{r features}
# annotate gene features
cov <- xfun::cache_rds({
  
  # load rda to get ref_cds object from DNDSCV
  refcds_38 <- '/lustre/scratch125/casm/team268im/fa8/119/DNDSCV_COVARIATES/refcds_GRCh38-GencodeV18+Appris.rda'
  load(refcds_38)
  
  # set gene names
  ref_cds <-
    RefCDS %>%
    setNames(RefCDS %>% purrr::map(~ .x$gene_name) %>% unlist())
  
  # annotate exons
  exons <-
    ref_cds[c(drivers$gene)] %>%
    purrr::map(~ .x$intervals_cds %>% 
                 tibble::as_tibble() %>%
                 dplyr::transmute(chr = .x$chr, start = V1, end = V2) %>%
                 dplyr::mutate(exon = paste0('exon', dplyr::row_number()))) %>%
    dplyr::bind_rows(.id = 'gene') %>%
    dplyr::mutate(pos = purrr::map2(start, end, ~ .x:.y)) %>%
    tidyr::unnest(cols = 'pos') %>%
    dplyr::select(-start, -end)
  cov <-
    cov %>%
    dplyr::left_join(exons, by = c('chr', 'pos', 'gene')) %>%
    dplyr::mutate(exon = ifelse(is.na(exon), 'intron', exon),
                  feature = ifelse(exon == 'intron', 'intron', 'exon'))
  
  # return
  cov
  
}, file = 'cov_w_features.rds', rerun = params$rerun)
```

```{r plots}
# get exon coords for plotting
exon_coords <- xfun::cache_rds({
  exon_coords <-
    cov %>%
    dplyr::filter(feature == 'exon') %>%
    dplyr::group_by(gene, exon) %>%
    dplyr::summarise(start = min(pos), 
                     end = max(pos))
  exon_coords
}, file = 'exon_coords.rds', rerun = params$rerun)

# function: plot all
plot_all <- function(p_df, p_exon_coords, p_title) {
  ggplot() +
    geom_rect(
      data = p_exon_coords,
      aes(xmin = start, xmax = end, ymin = -Inf, ymax = Inf),
      colour = 'lightblue', fill = 'lightblue'
    ) +
    geom_col(
      data = p_df,
      aes(x = pos, y = depth, fill = feature)) +
    theme_classic() +
    theme(panel.border = element_rect(fill = NA),
          panel.spacing = unit(.05, 'lines')) +
    scale_fill_manual(values = c(exon = 'darkblue', intron = 'darkgrey')) +
    labs(title = paste(p_title, '(all)')) +
    facet_grid(celltype ~ ., space = 'free_x', scales = 'free_x') 
}

# function: plot exons 
plot_exons <- function(p_df, p_title) {
  p_df %>%
    # get only exonic sequence
    dplyr::filter(feature == 'exon') %>%
    # fix exon ordering
    dplyr::mutate(
      exon = exon %>% forcats::fct_reorder(exon %>% gsub('exon', '', .) %>% as.numeric())
    ) %>%
    ggplot(aes(x = pos, y = depth, group = celltype)) +
    geom_col(fill = 'darkblue') +
    theme_classic() +
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          panel.spacing = unit(.05, 'lines'),
          panel.border = element_rect(fill = NA)) +
    labs(title = paste(p_title, '(exons)')) +
    facet_grid(celltype ~ exon, space = 'free_x', scales = 'free_x') 
}

# create list of plots by gene x sample
p <- xfun::cache_rds({
  
  # initiate plot list
  p <- list()
  n_cts <- list()
  
  cov %>% 
    dplyr::group_by(gene, sample) %>%
    dplyr::group_split() %>%
    purrr::map(function(df_i) {
      
      # for testing:
      # cov %>% dplyr::group_by(gene, sample) %>% dplyr::group_split() %>% {.[[1]]} -> df_i
      
      # get info
      i <- 
        df_i %>%
        dplyr::distinct(gene, sample)
      info <- t(i[1,]) %>% paste(collapse = '.')
      p[[info]] <<- list()
      
      # get number of celltypes for plot dims
      p[[info]]$n_cts <<- 
        df_i %>% 
        dplyr::filter(!is.na(celltype)) %>% 
        dplyr::pull(celltype) %>% 
        dplyr::n_distinct()
      
      # plot by sample and by celltype
      list(by_sample = list(lvl = 'drivers_samtools_depth', title_suffix = ''),
           by_celltype = list(lvl = 'drivers_samtools_depth_by_celltype', title_suffix = '_by_celltype')) %>%
        purrr::map(function(x){
          
          # plot all
          p[[info]][[paste0('all', x$title_suffix)]] <<-
            plot_all(
              p_df = df_i %>% dplyr::filter(lvl == x$lvl),
              p_exon_coords = exon_coords %>% dplyr::filter(gene == i$gene),
              p_title = paste0(info, x$title_suffix)
            )  
          
          # plot exons only
          p[[info]][[paste0('exons', x$title_suffix)]] <<-
            plot_exons(
              p_df = df_i %>% dplyr::filter(lvl == x$lvl),
              p_title = paste0(info, x$title_suffix)
            )
          
        })
      
    })
  
  # return 
  p
  
}, file = 'plots.rds', rerun = params$rerun)
```

```{r knit_plots, echo = F, warning = F, message = F, results = 'asis'}
chunks <- c()

for (info in names(p)) {
  chunks <- c(chunks, paste0('\n####', info))
  message(info)
  
  knit_plot <- c(
    '',
    paste0('```{r ', info, ', class.source = "fold-hide", warning = F, message = F, ',
           'fig.dim = c(20, ', 4 + 4 * p[[info]]$n_cts, ')}'),
    paste0('p_i <- p[["', info, '"]]'),
    paste0('grid.arrange(p_i$all, p_i$all_by_celltype, p_i$exons, p_i$exons_by_celltype, ncol = 1, ',
           'heights = c(2, p_i$n_cts * 2, 2, p_i$n_cts * 2))'),
    '```',
    '')
  knitted_plot <- knitr::knit_child(text = knit_plot, envir = environment(), quiet = T)
  dev_off_if()
  
  chunks <- c(chunks, knitted_plot)
}
dev_off_if()

# evaluate chunks
cat(unlist(chunks), sep = '\n')
```

```{r mean_coverage}
cov %>%
  dplyr::filter(lvl == 'drivers_samtools_depth') %>%
  dplyr::group_by(sample, gene, feature) %>%
  dplyr::summarise(mean_depth = mean(depth)) %>%
  ggplot(aes(x = sample, y = mean_depth, fill = feature)) +
  geom_col(position = 'dodge') +
  facet_grid(gene ~ .) +
  theme_classic()
```