---
title: "Targeted interrogation of mutational hotspots in rheumatoid arthritis"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: true
    code_folding: show
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
params:
  recursites: NULL
  cache_dir: './target_mutation_calling_cache/'
  rerun: true
---

```{r setup, include = F, message = F, warning = F, class.source = 'fold-hide'}
# wd <- ifelse(Sys.info()['nodename'] == "mib119104s", '~/Volumes/', '') |> paste0('/lustre/scratch125/casm/team268im/at31/RA_som_mut/scomatic/') ; setwd(wd)

# chunk options
knitr::opts_chunk$set(
  warning = FALSE,
  message = FALSE,
  dpi = 300, 
  cache.path = params$cache_dir) 

# libraries
library(magrittr)
library(ggplot2)

# dev.off() if there is a device
dev_off_if <- function() {if (length(dev.list()) != 0) { dev.off() }}
```

```{r load_files}
# all possible nucleotides
nucleotides <- c("A", "T", "C", "G", "-", "INS", "a", "t", "c", "g", "_", "ins")

# strandless nucleotides - colours and order
nt_cols <-
  c('INS' = '#e6ab02',
    'DEL' = '#1c2d78',
    'A' = '#1b9e77', 
    'C' = '#d95f02', 
    'G' = '#7570b3', 
    'T' = '#e7298a',
    'ref' = 'grey')

# recursites (to add detected mut)
recursites <-
  readr::read_tsv(params$recursites) 
recurmuts <-
  recursites %>%
  dplyr::transmute(
    chr,
    mut_pos = pos,
    mut_pos_lab = paste0(pos, '\ mut: ', mut)
  )

# reads
reads <-
  list.files(pattern = '_reads.tsv') %>%
  purrr::set_names(., gsub('_reads.tsv', '', .)) %>%
  purrr::map(function(file) {
    readr::read_tsv(file, show_col_types = F)
  }) %>%
  dplyr::bind_rows(.id = 'id') %>%
  tidyr::pivot_longer(cols = dplyr::all_of(nucleotides)) %>%
  dplyr::mutate(
    # colour in alt alleles
    status = dplyr::case_when(toupper(name) == ref ~ 'ref',
                              name %in% c('-', '_') ~ 'DEL',
                              TRUE ~ toupper(name)) %>%
      forcats::fct_relevel(names(nt_cols))) %>%
  # add detected mutations
  dplyr::left_join(recurmuts, by = c('chr', 'mut_pos')) %>%
  split(., .$gene)

# save reads
reads %>% 
  saveRDS('reads.rds')
```

## Recurrent mutations sites

Recurrent mutations from the LCM-WES data were found using `dNdScv`. Mutated sites in the following genes will be interrogated.

```{r plot_recurmuts, fig.width = 10, eval = F}
recursites %>%
  dplyr::group_by(gene) %>%
  dplyr::mutate(sum_freq = sum(freq)) %>%
  ggplot(aes(x = reorder(gene, -sum_freq), y = freq, fill = qval)) +
  geom_col(position = 'stack') +
  scale_x_discrete(guide = guide_axis(angle = -90)) +
  labs(x = 'gene') +
  theme_classic()
```

## Mutations

First, we plot the rates of reference and non-reference alleles at mutation windows in all genes. Then, we do the same for substitutions only (ignoring indels).

```{r plot}
# function: create plot list
plot_positions <- function(reads, include_indels = T, include_refs = T) {
  p <-
    names(reads) %>%
    purrr::set_names() %>%
    purrr::map(function(gene) {
      
      p_dat <-
        reads[[gene]] %>%
        # optionally exclude indels
        dplyr::filter(
          include_indels == T | !status %in% c('DEL', 'INS'),
          include_refs == T | status != 'ref') %>%
        # only plot celltypes with any reads
        dplyr::group_by(celltype) %>%
        dplyr::filter(max(value) > 0) %>%
        dplyr::ungroup() 
      
      if (nrow(p_dat) > 0) {
        
        # get ref alleles for x axis labels
        x_labs <-
          p_dat %>%
          dplyr::distinct(pos, ref) 
        x_labs_vec <- x_labs$ref
        names(x_labs_vec) <- x_labs$pos
        
        g_title <- paste0(
          gene, ', w/', 
          ifelse(include_indels == T, '', 'o'), 
          ' indels',
          ifelse(include_refs == T, '', ', w/o refs'))
        
        # get plot dims
        g_p_dims <-
          c(min(max(7, (dplyr::n_distinct(p_dat$mut_pos) + 2) * 2), 50),
            min(max(7, (dplyr::n_distinct(p_dat$celltype)) * 1.5), 50))
        
        # get plot
        g_p <-
          p_dat %>%
          dplyr::mutate(
            # make sure ordering of positions is maintained
            x_pos = factor(as.character(pos), 
                           levels = unique(as.character(sort(pos)))),
            x_mut_pos = dplyr::case_when(pos == mut_pos ~ x_pos, TRUE ~ NA)) %>%
          # plot
          ggplot(aes(x = x_pos, y = value, fill = status)) +
          geom_col(position = 'stack') +
          geom_vline(aes(xintercept = x_mut_pos)) +
          theme_classic() +
          scale_x_discrete(labels = x_labs_vec) +
          scale_fill_manual(values = nt_cols) +
          ggh4x::facet_grid2(celltype ~ mut_pos_lab, scales = 'free') +
          labs(title = g_title, x = 'pos (ref shown)')
        
        # return
        list(plot = g_p, dims = g_p_dims)
        
      }
    })
  
  # return
  p
}

# generate plot lists
p_w_indels <- plot_positions(reads)
p_wo_indels <- plot_positions(reads, F)
p_wo_indels_wo_refs <- plot_positions(reads, F, F)
```

```{r, results = 'asis'}
chunks <- c()

for (gene in names(reads)) {
  
  chunks <- c(chunks, paste0('\n### ', gene))
  
  knit_plot <- c(
    '',
    paste0('```{r ', gene,'_w_indels, fig.dim = p_w_indels[["', gene, '"]]$dims, warning = F, message = F, echo = F}'),
    paste0('print(p_w_indels[["', gene, '"]]$plot)'),
    '```',
    '',
    paste0('```{r ', gene,'_wo_indels, fig.dim = p_wo_indels[["', gene, '"]]$dims, warning = F, message = F, echo = F}'),
    paste0('print(p_wo_indels[["', gene, '"]]$plot)'),
    '```',
    '',
    paste0('```{r ', gene,'_wo_indels_wo_refs, fig.dim = p_wo_indels_wo_refs[["', gene, '"]]$dims, warning = F, message = F, echo = F}'),
    paste0('print(p_wo_indels_wo_refs[["', gene, '"]]$plot)'),
    '```',
    ''
  )
  
  knitted_plot <- knitr::knit_child(text = knit_plot, envir = environment(), quiet = T)
  dev_off_if()
  
  chunks <- c(chunks, knitted_plot)
  
}

cat(unlist(chunks), sep = '\n')
```

## Mutation types, by patient across genes

### Mutation types, by patients across all genes

```{r muts_by_patient, fig.width = 10, class.source = 'fold-hide'}
reads %>%
  dplyr::bind_rows() %>%
  ggplot(aes(x = id, y = value, fill = status)) +
  geom_col() +
  scale_fill_manual(values = nt_cols) + 
  theme_classic() +
  scale_x_discrete(guide = guide_axis(angle = -90))
```

### Mutation types, by patient across all genes, without indels

```{r muts_by_patient_no_indels, fig.width = 10, class.source = 'fold-hide'}
reads %>%
  dplyr::bind_rows() %>%
  dplyr::filter(!status %in% c('INS', 'DEL', 'ref')) %>%
  ggplot(aes(x = id, y = value, fill = status)) +
  geom_col() +
  scale_fill_manual(values = nt_cols) + 
  theme_classic() +
  scale_x_discrete(guide = guide_axis(angle = -90))
```
