---
title: "Targeted interrogation of mutational hotspots in rheumatoid arthritis"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: false
    code_folding: show
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
params:
  drivers: NULL
  hotspots: NULL
  bam_mappings: NULL
  cache_dir: './target_mutation_calling_cache/'
  rerun: true
---

```{r setup, include = F, message = F, warning = F, class.source = 'fold-hide'}
# wd <- ifelse(Sys.info()['nodename'] == "mib119104s", '~/Volumes/', '') |> paste0('/lustre/scratch125/casm/team268im/at31/RA_som_mut/scomatic/') ; setwd(wd)
# rmarkdown::render('reports/driver_coverage.Rmd', output_file = 'driver_coverage.html', output_dir = 'reports/', params = list(rerun = F))

# chunk options
knitr::opts_knit$set(root.dir = '../')
knitr::opts_chunk$set(
  warning = FALSE,
  message = FALSE,
  dpi = 300, 
  cache.path = params$cache_dir) 

# libraries
library(magrittr)
library(ggplot2)
library(GenomicRanges)

# dev.off() if there is a device
dev_off_if <- function() {if (length(dev.list()) != 0) { dev.off() }}
```

```{r load_files}
list.files(pattern = '_reads.tsv') %>%
  purrr::set_names(., gsub('_reads.tsv', '', .)) %>%
  purrr::map(function(file) {
    reads <- readr::read_tsv(reads)
  })
```

```{r plot, fig.width = 20}
# plot positions
unique(sites$gene) %>%
  purrr::walk(function(i) {
    p <-
      sites %>%
      dplyr::filter(gene == i) %>%
      dplyr::left_join(refs, by = c('chr', 'pos')) %>%
      tidyr::pivot_longer(cols = dplyr::all_of(nucleotides),
                          values_to = 'nt_n_reads', names_to = 'nt') %>%
      dplyr::mutate(
        strandless_nt = 
          ifelse(nt %in% c('-', '_'), 'DEL', toupper(nt)),
        is_ref = ifelse(strandless_nt == ref, 'ref', 'alt'),
        strandless_nt = paste0(strandless_nt, '_', is_ref) %>% 
          forcats::fct_relevel(unique(c(paste0(toupper(nucleotides), '_ref'),
                                        paste0(toupper(nucleotides), '_alt'))))) %>%
      ggplot(aes(x = pos, y = nt_n_reads, fill = strandless_nt)) + 
      geom_vline(aes(xintercept = mut_pos)) +
      geom_col(position = 'stack') +
      theme_bw() +
      theme(axis.text.x = element_blank()) +
      ggh4x::facet_grid2(celltype ~ mut_pos, scales = 'free') +
      scale_fill_manual(values = c(A_ref = 'darkgreen', A_alt = 'lightgreen',
                                   C_ref = 'darkblue', C_alt = 'lightblue',
                                   G_ref = 'darkorange3', G_alt = 'peachpuff',
                                   T_ref = 'darkred', T_alt = 'salmon',
                                   INS_alt = 'black', DEL_alt = 'grey')) +
      ggtitle(i)
    print(p)
    })
```

# Mutational calls in drivers from the LCM-WES data

```{r mut_hotspots, eval = F}
# read in bam mappings, get celltype-split bams only
mappings <-
  readr::read_csv(params$bam_mappings) %>%
  dplyr::filter(!is.na(celltype)) %>%
  # temporarily remove problematic BRI-401 endo bam
  dplyr::filter(!(id == 'BRI_401' & celltype == 'Endothelial_cell'))

# read in drivers
drivers <-
  readr::read_tsv(params$drivers, col_names = c('chr', 'start', 'end', 'gene')) %>%
  dplyr::mutate(chr = gsub('chr', '', chr))

# read in mutations in drivers
hotspots <-
  readr::read_tsv(params$hotspots) %>%
  dplyr::mutate(mut_pos = pos,
                chr = gsub('chr', '', chr)) %>%
  dplyr::group_by(chr, mut_pos, gene, ref_cod) %>%
  # summarise each position
  dplyr::summarise(n = dplyr::n(),
                   id = paste(id, collapse = ','),
                   mut_cod = paste(mut_cod, collapse = ','),
                   impact = paste(impact, collapse = ',')) %>% 
  dplyr::group_by(dplyr::across(everything())) %>%
  # extend pileup window 20bp either side of mutated position
  dplyr::reframe(pos = (mut_pos - 20):(mut_pos + 20)) 

# save as a bed
hotspots %>%
  dplyr::transmute(chr = paste0('chr', chr), start = pos - 1, end = pos) %>%
  readr::write_tsv(
    'data/driver_genes/lcm_wes_mutations_20bp_window.bed',
    col_names = F)

# get ref allele at each position
refs <- xfun::cache_rds({
  refs <-
    system('bedtools getfasta -fi /lustre/scratch125/casm/team268im/al28/bed_ref/GRCh38_full_analysis_set_plus_decoy_hla_genome.fa -bed data/driver_genes/lcm_wes_mutations_20bp_window.bed -tab', intern = T) %>%
    paste(collapse = '\n') %>%
    readr::read_tsv(col_names = c('pos', 'ref')) %>%
    tidyr::separate_wider_delim(cols = pos, delim = stringr::regex(':|-'), 
                                names = c('chr', 'start', 'end')) %>%
    type.convert(as.is = T) %>%
    dplyr::transmute(chr = gsub('chr', '', chr), pos = end, ref) %>%
    dplyr::distinct()
  refs
}, file = 'refs.rds', rerun = params$rerun)
```


