---
title: "Gains and losses in rheumatoid arthritis data from inferCNV"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: false
    code_folding: hide
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
params:
  infercnv_dir: NULL
  cache_dir: NULL
  rerun: true
---

```{r setup, include = F, message = F, warning = F, class.source = 'fold-hide'}
# wd <- ifelse(Sys.info()['nodename'] == "mib119104s", '~/Volumes/', '') |> paste0('/lustre/scratch125/casm/team268im/at31/RA_som_mut/infercnv/') ; setwd(wd)
# rmarkdown::render('reports/infercnv.Rmd', output_file = 'driver_coverage.html', output_dir = 'reports/', params = list(rerun = T))

# libraries
library(magrittr)
library(ggplot2)
library(gridExtra)
library(Matrix)
library(SummarizedExperiment)
library(SingleCellExperiment)

# dev.off() if there is a device
dev_off_if <- function() {if (length(dev.list()) != 0) { dev.off() }}

# chunk options
knitr::opts_knit$set(root.dir = '../')
knitr::opts_chunk$set(
  warning = FALSE, 
  dpi = 300, 
  message = FALSE,
  cache.path = params$cache_dir) 
```

inferCNV was run on the scRNA-seq data to call copy number variants in synovial cells. No reference group was used, so the average profile of all cells was used in each patient as the baseline. 

First, we load the CITE-seq object. We convert it to a SingleCellExperiment object. Next, we add the CNV metadata from the inferCNV run to the object.

```{r load_obj, results = 'asis'}
# dirs
data_dir <- '/lustre/scratch126/casm/team268im/mp34/scRNAseq_data/RA_ZhangEtal2023/processed_output/'
infercnv_dir <- 'out/Zhang2023/by_celltype/'

# convert to sce
sce <- xfun::cache_rds({
  
  # load object
  obj <- 
    readRDS(file.path(data_dir, 'all_cells_reference.rds'))
  
  # load rna data
  assays <-
    list(
      counts = 
        readRDS(file.path(data_dir, 'raw_mRNA_count_matrix.rds')),
      logcounts = 
        readRDS(file.path(data_dir, 'qc_mRNA_314011cells_log_normalized_matrix.rds')))
  
  # get cell order from counts
  cell_order <- colnames(assays$logcounts)
  
  # filter and reorder to cells that pass qc from counts
  assays$counts <- 
    assays$counts[rownames(assays$logcounts), 
                  cell_order]
  
  # generate coldata
  col_data <-
    obj$meta_data[cell_order, ]
  
  # populate sce
  sce <- 
    SingleCellExperiment::SingleCellExperiment(
      assays = assays,
      colData = col_data
    )
  
  # name, reorder, and add umap embeddings
  umap <- obj$umap$embedding
  rownames(umap) <- rownames(obj$meta_data)
  umap <- umap[cell_order, ]
  reducedDim(sce, 'UMAP_uwot') <- umap
  
  # get cnv metadata
  cnvs <- 
    list.files(
      infercnv_dir,
      pattern = 'map_metadata_from_infercnv.txt',
      recursive = T) %>%
    purrr::set_names(., dirname(.)) %>%
    purrr::map(function(file) {
      file.path(infercnv_dir, file) %>% 
        read.table(sep = '\t')
      }) %>%
    dplyr::bind_rows() 
  
  # left join cnvs to coldata, reorder, restore rownames
  col_data <- merge(col_data, cnvs, by = 'row.names', all.x = T)
  rownames(col_data) <- col_data[, 1]
  col_data <- 
    col_data[colnames(sce), -1] %>%
    DataFrame()
  
  # add to sce
  sce@colData <- col_data
  
  # return
  sce

}, file = 'sce.rds')
```

The `inferCNV` output file `map_metadata_from_infercnv.txt` contains columns with different CNV measures at whole-chromosome level per cell, with rownames denoting cells. The columns have the following values...

```{r echo = F, results = 'asis'}
tibble::tribble(
  ~name, ~`type - value`,
  '`has_*`', 'binary - whether any CNV / loss / duplication is found within the given chr',
  '`proportion_*`', 'proportion - proportion of genes that are part of any CNV / loss / duplication within the given chr',
  '`proportion_scaled_*`', 'proportion - same as proportion_*, but taking into account whether genes that are duplicated / lost are a single copy or two copies, as a weight',
  '`top_loss_*`', 'the top CNVs that affect the most genes'
) %>%
  knitr::kable()
```

First, we plot UMAPs by ID and celltype.

```{r plot_umaps, fig.height = 10}
dittoSeq::dittoDimPlot(sce, 'cell_type', size = 0.2) +
  coord_fixed()
```

```{r plot_umaps_by_id, fig.height = 30}
dittoSeq::dittoDimPlot(sce, 'cell_type', size = 0.2, 
                       split.by = 'sample') +
  coord_fixed()
```

Then, we visualise chr7 and chr5 gains.

```{r plot_chr_gains}
dittoSeq::dittoDimPlot(
  sce, 'proportion_dupli_chr7', size = 0.2) +
  coord_fixed()
dittoSeq::dittoDimPlot(
  sce, 'proportion_dupli_chr7', size = 0.2, split.by = 'sample') +
  coord_fixed()
dittoSeq::dittoDimPlot(
  sce, 'proportion_dupli_chr5', size = 0.2) +
  coord_fixed()
dittoSeq::dittoDimPlot(
  sce, 'proportion_dupli_chr5', size = 0.2, split.by = 'sample') +
  coord_fixed()
```

Now, we will plot gain and loss events by chromosome.

```{r plot_gains_and_losses, fig.height = 20}
# mean proportion loss / gain per chromosome per celltype per id
p_dat <-
  sce@colData %>%
  tibble::as_tibble() %>%
  dplyr::group_by(sample, cell_type) %>%
  dplyr::summarise(
    dplyr::across(dplyr::starts_with('proportion_'), ~ mean(.x, na.rm = T))) %>%
  dplyr::ungroup()

# split gain and loss
p_dat <-
  c('dupli', 'loss') %>% 
  purrr::set_names() %>%
  purrr::map(function(direction) {
    p_dat_i <-
      p_dat %>% 
        dplyr::filter(!is.na(paste0('proportion_', direction, '_chr1')))
    list(
      matrix = 
        p_dat_i %>%
        dplyr::select(dplyr::contains(paste0('proportion_', direction))) %>%
        as.matrix(),
      annot = 
        p_dat_i %>% 
        dplyr::select(sample, cell_type)
    )
  })

# plot heatmaps
pheatmap::pheatmap(p_dat$dupli$matrix, cluster_cols = F, cluster_rows = F)
pheatmap::pheatmap(sce@colData[,grepl('proportion_dupli', colnames(sce@colData))])
```

#### chr7 gain

We are most interested in the chr7 gain events in fibroblasts. We visualise their distribution across the dataset. 

```{r chr7_prop_dist}
p_dat <-
  sce@colData %>%
  tibble::as_tibble() %>%
  dplyr::filter(!is.na(proportion_dupli_chr7)) 

plot_prop_chr7_gain <- function(p_dat, position_arg) {
  p_dat %>%
    ggplot() +
    geom_bar(aes(x = cell_type, 
                 fill = proportion_dupli_chr7, 
                 group = proportion_dupli_chr7),
             position = position_arg) +
    scale_x_discrete(guide = guide_axis(angle = -45)) +
    theme_classic() +
    scale_fill_viridis_b(n.breaks = 9) 
}

plot_prop_chr7_gain(p_dat, position_stack(reverse = T)) +
  ggtitle('chr7 gain count - all cells')
plot_prop_chr7_gain(p_dat, position_fill(reverse = T)) +
  ggtitle('chr7 gain proportion - all cells')
p_dat %>%
  {split(., .$id)} %>%
  purrr::map2(names(.), ., function(id, df) {
    plot_prop_chr7_gain(df, position_fill(reverse = T)) +
      ggtitle(paste0('chr7 gain count - ', id))
    plot_prop_chr7_gain(df, position_stack(reverse = T)) +
      ggtitle(paste0('chr7 gain proportion - ', id))
  })
```

##### Pseudobulk

We pseudobulk chr7+ cells and perform differential expression analysis. In order to do this, we use the `glmGamPoi` package in R. This analysis follows [this tutorial](https://bioconductor.org/packages/devel/bioc/vignettes/glmGamPoi/inst/doc/pseudobulk.html).

We will define cells with chr7 gain as those affecting >90% of genes. We will pseudobulk by patient, celltype, and chr7 ploidy to produce a reduced object.

```{r pseudobulk_chr7_gain}
reduced_sce <- xfun::cache_rds({
  
  # we assign chr7 status as >90% of genes affected by a duplication
  sce@colData$chr7_gain <-
    as.numeric(sce@colData$proportion_dupli_chr7 > 0.9)
  
  # pseudobulk the sce object by patient, celltype, and chr7 ploidy
  reduced_sce <-
    glmGamPoi::pseudobulk_sce(
      sce, 
      group_by = vars(sample, cell_type, chr7_gain),
      n_cells = n()
    )
  
  # remove NAs
  reduced_sce <- reduced_sce[, !is.na(reduced_sce$cell_type)]
  
  # return
  reduced_sce
  
}, file = 'reduced_sce.rds')
```

With the reduced data, we can conduct differential expression analysis, treating the data like bulk RNA-seq data. 

```{r deseq, message = T}
# use DESeq2's size factor calculation procedure
fit <-
  glmGamPoi::glm_gp(
    reduced_sce,
    design = ~ chr7_gain * cell_type + sample,
    size_factor = 'ratio',
    verbose = T
  )
res <-
  glmGamPoi::test_de(
    fit, 
    contrast = 
      cond(cell_type = 'Stromal cell', chr7_gain = 1) -
      cond(cell_type = 'Stromal cell', chr7_gain = 0)
  )
```

```{r plot_de}
res %>%
  ggplot(aes(x = lfc, y = log10(pval))) +
  geom_point(aes(colour = adj_pval < 0.01), size = 0.5) +
  t
```
