---
title: "Gains and losses in rheumatoid arthritis data from inferCNV"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: false
    code_folding: hide
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
params:
  infercnv_dir: NULL
  rerun: true
---

```{r setup, include = F, message = F, warning = F, class.source = 'fold-hide'}
# wd <- ifelse(Sys.info()['nodename'] == "mib119104s", '~/Volumes/', '') |> paste0('/lustre/scratch125/casm/team268im/at31/RA_som_mut/infercnv/') ; setwd(wd)
# rmarkdown::render('reports/infercnv.Rmd', output_file = 'driver_coverage.html', output_dir = 'reports/', params = list(rerun = T))

# chunk options
knitr::opts_knit$set(root.dir = '../')
knitr::opts_chunk$set(
  warning = FALSE, 
  dpi = 300, 
  message = FALSE) 

# libraries
library(magrittr)
library(ggplot2)
library(gridExtra)

# dev.off() if there is a device
dev_off_if <- function() {if (length(dev.list()) != 0) { dev.off() }}
```

inferCNV was run on the scRNA-seq data to call copy number variants in synovial cells. No reference group was used, so the average profile of all cells was used in each patient as the baseline. 

First, we load the CITE-seq object and extract the metadata and UMAP embeddings.

```{r load_obj, results = 'asis'}
# infercnv output dir
infercnv_dir <- 'out/Zhang2023/by_celltype/'

# load object
obj <- 
  readRDS('/lustre/scratch126/casm/team268im/mp34/scRNAseq_data/RA_ZhangEtal2023/processed_output/all_cells_reference.rds')

# get umap embeddings
umap <-
  obj$umap$embedding %>%
  tibble::as_tibble() %>%
  dplyr::bind_cols(
    # add metadata
    obj$meta_data %>%
      tibble::as_tibble() %>%
      dplyr::mutate(id = sample)) 
```

Next, we add the CNV metadata from the inferCNV run. We must wrangle it from wide- to long-format.

```{r infercnv_output}
# get cnv metadata
cnvs <- 
  list.files(
    infercnv_dir,
    pattern = 'map_metadata_from_infercnv.txt',
    recursive = T) %>%
  purrr::set_names(., dirname(.)) %>%
  purrr::map(function(file) {
    file.path(infercnv_dir, file) %>% 
      read.table(sep = '\t') %>% 
      tibble::as_tibble(rownames = 'cell')
    }) %>%
  dplyr::bind_rows(.id = 'id') %>%
  # convert from wide to long format
  tidyr::pivot_longer(
    cols = tidyselect::starts_with(c('has_', 'proportion_', 'top_'))
  ) %>%
  dplyr::rowwise() %>%
  dplyr::mutate(
    chr = name %>%
      gsub('.*_', '', .) %>%
      {forcats::fct_reorder(., gsub('chr', '', .) %>% as.numeric())},
    event = sub(paste0(chr, '_'), '', name),
    measure = sub(paste0(chr, '_', event), '', name)) %>%
  dplyr::ungroup()

# combine umap and cnvs
umap <-
  umap %>%
  # add cnv metadata
  dplyr::inner_join(cnvs, by = c('id', 'cell')) 
```

The `inferCNV` output file `map_metadata_from_infercnv.txt` contains columns with different CNV measures at whole-chromosome level per cell, with rownames denoting cells. The columns have the following values...

```{r echo = F, results = 'asis'}
tibble::tribble(
  ~name, ~`type - value`,
  '`has_*`', 'binary - whether any CNV / loss / duplication is found within the given chr',
  '`proportion_*`', 'proportion - proportion of genes that are part of any CNV / loss / duplication within the given chr',
  '`proportion_scaled_*`', 'proportion - same as proportion_*, but taking into account whether genes that are duplicated / lost are a single copy or two copies, as a weight',
  '`top_loss_*`', 'the top CNVs that affect the most genes'
) %>%
  knitr::kable()
```

We will define whole chromosome events as those affecting >90% of genes in a given chromosome.

```{r call_ploidy}

```

Now, we will plot gain and loss events by chromosome

```{r}
# function: plot gains and losses
plot_gains_and_losses <- 
  function(umap, annot) {
    umap %>%
      dplyr::rename(annotation = tidyselect::all_of(annot)) %>%
      dplyr::filter(measure == 'has', event != 'cnv') %>%
      dplyr::group_by(id, chr, event, annotation) %>%
      dplyr::summarise(n_cells = dplyr::n(),
                       n_events = sum(value),
                       prop_of_cells = n_events / n_cells) %>%
      ggplot(aes(x = id, y = chr, alpha = prop_of_cells, fill = event)) +
      geom_tile() +
      theme_classic() +
      theme(panel.spacing = unit(0, "lines")) +
      scale_fill_manual(values = c('loss' = scales::muted('blue'),
                                   'dupli' = scales::muted('red'))) +
      scale_alpha(range = c(0, 1)) +
      facet_grid(event ~ annotation)
  }

plot_gains_and_losses(umap, 'cell_type')
plot_gains_and_losses(umap, 'cluster_name')

# plot chr7 gain on umap
umap %>%
  dplyr::filter(measure == 'has', event == 'dupli', chr == 'chr7') %>%
  ggplot(aes(x = UMAP1, y = UMAP2, colour = cell_type, alpha = has_dupli_chr7)) +
  geom_point(size = 0.1) +
  theme_classic() +
  coord_fixed()
```

