---
title: "Gains and losses in rheumatoid arthritis data from inferCNV"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: false
    code_folding: hide
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
params:
  infercnv_dir: NULL
  cache_dir: NULL
  rerun: true
---

```{r setup, include = F, message = F, warning = F, class.source = 'fold-hide'}
# wd <- ifelse(Sys.info()['nodename'] == "mib119104s", '~/Volumes/', '') |> paste0('/lustre/scratch125/casm/team268im/at31/RA_som_mut/infercnv/') ; setwd(wd)
# rmarkdown::render('reports/infercnv.Rmd', output_file = 'driver_coverage.html', output_dir = 'reports/', params = list(rerun = T))

# libraries
library(magrittr)
library(ggplot2)
library(gridExtra)
library(Matrix)
library(SummarizedExperiment)
library(SingleCellExperiment)

# dev.off() if there is a device
dev_off_if <- function() {if (length(dev.list()) != 0) { dev.off() }}

# chunk options
knitr::opts_knit$set(root.dir = '../')
knitr::opts_chunk$set(
  warning = FALSE, 
  dpi = 300, 
  message = FALSE,
  cache.path = params$cache_dir) 
```

inferCNV was run on the scRNA-seq data to call copy number variants in synovial cells. No reference group was used, so the average profile of all cells was used in each patient as the baseline. 

First, we load the CITE-seq object. We convert it to a SingleCellExperiment. 

```{r load_obj, results = 'asis'}
# dirs
data_dir <- '/lustre/scratch126/casm/team268im/mp34/scRNAseq_data/RA_ZhangEtal2023/processed_output/'
infercnv_dir <- 'out/Zhang2023/by_celltype/'

# convert to sce
sce <- xfun::cache_rds({
  
  # load object
  obj <- 
    readRDS(file.path(data_dir, 'all_cells_reference.rds'))
  
  # load rna data
  assays <-
    list(
      counts = 
        readRDS(file.path(data_dir, 'raw_mRNA_count_matrix.rds')),
      logcounts = 
        readRDS(file.path(data_dir, 'qc_mRNA_314011cells_log_normalized_matrix.rds')))
  
  # get cell order from counts
  cell_order <- colnames(assays$logcounts)
  
  # filter and reorder to cells that pass qc from counts
  assays$counts <- 
    assays$counts[rownames(assays$logcounts), 
                  cell_order]
  
  # generate coldata
  colData <-
    obj$meta_data[cell_order, ]
  
  # populate sce
  sce <- 
    SingleCellExperiment::SingleCellExperiment(
      assays = assays,
      colData = colData
    )
  
  # name, reorder, and add umap embeddings
  umap <- obj$umap$embedding
  rownames(umap) <- rownames(obj$meta_data)
  umap <- umap[cell_order, ]
  reducedDim(sce, 'UMAP_uwot') <- umap
  
  # return
  sce

}, file = 'sce.rds')

sce <- xfun::cache_rds({
  
  # get cnv metadata, reorder, left join to colData
  cnvs <- 
    list.files(
      infercnv_dir,
      pattern = 'map_metadata_from_infercnv.txt',
      recursive = T) %>%
    purrr::set_names(., dirname(.)) %>%
    purrr::map(function(file) {
      file.path(infercnv_dir, file) %>% 
        read.table(sep = '\t')
      }) %>%
    dplyr::bind_rows() 
  
  # left join cnvs to coldata, reorder, restore rownames
  colData <- merge(colData, cnvs, by = 'row.names', all.x = T)
  rownames(colData) <- colData[, 1]
  colData <- colData[, -1]
  colData <- colData[colnames(sce), ]
  
  # add to sce
  sce@colData <- colData
  
  # return
  sce
  
}, file = 'sce_w_infercnv.rds')
```

Next, we add the CNV metadata from the inferCNV run to the UMAP embeddings. We must wrangle it from wide- to long-format.

```{r infercnv_output}
cnvs <- xfun::cache_rds({
  
  # get cnv metadata
  cnvs <- 
    list.files(
      infercnv_dir,
      pattern = 'map_metadata_from_infercnv.txt',
      recursive = T) %>%
    purrr::set_names(., dirname(.)) %>%
    purrr::map(function(file) {
      file.path(infercnv_dir, file) %>% 
        read.table(sep = '\t') %>% 
        tibble::as_tibble(rownames = 'cell')
      }) %>%
    dplyr::bind_rows(.id = 'id') %>%
    # convert from wide to long format
    tidyr::pivot_longer(
      cols = tidyselect::starts_with(c('has_', 'proportion_', 'top_'))
    ) 
  
  # get chr / event / name
  names <-
    cnvs %>%
    dplyr::distinct(name) %>%
    dplyr::rowwise() %>%
    dplyr::mutate(
      chr = gsub('.*_', '', name),
      event = gsub(paste0('_', chr), '', name) %>% sub('.*_', '', .),
      measure = sub(paste0('_', event, '_', chr), '', name),
      chr = dplyr::case_when(
        measure == 'top' ~ paste0('chr', chr), TRUE ~ chr) %>% 
        {forcats::fct_reorder(., gsub('chr', '', .) %>% as.numeric())}) %>%
    dplyr::ungroup() 
  
  # combine umap, names, and cnvs
  cnvs <-
    cnvs %>%
    # add names
    dplyr::left_join(names, by = 'name')
  
  # return
  cnvs 
  
}, file = 'cnvs.rds')
```

The `inferCNV` output file `map_metadata_from_infercnv.txt` contains columns with different CNV measures at whole-chromosome level per cell, with rownames denoting cells. The columns have the following values...

```{r echo = F, results = 'asis'}
tibble::tribble(
  ~name, ~`type - value`,
  '`has_*`', 'binary - whether any CNV / loss / duplication is found within the given chr',
  '`proportion_*`', 'proportion - proportion of genes that are part of any CNV / loss / duplication within the given chr',
  '`proportion_scaled_*`', 'proportion - same as proportion_*, but taking into account whether genes that are duplicated / lost are a single copy or two copies, as a weight',
  '`top_loss_*`', 'the top CNVs that affect the most genes'
) %>%
  knitr::kable()
```

Now, we will plot gain and loss events by chromosome

```{r plot_gains_and_losses, fig.height = 20}
# function: plot gains and losses
plot_gains_and_losses <- 
  function(cnvs, annot) {
    cnvs %>%
      dplyr::rename(annotation = tidyselect::all_of(annot)) %>%
      dplyr::filter(measure == 'proportion', event != 'cnv') %>%
      dplyr::group_by(id, chr, event, annotation) %>%
      dplyr::summarise(proportion = mean(value)) %>%
      ggplot(aes(x = chr, y = id, alpha = proportion, fill = event)) +
      geom_tile() +
      theme_classic() +
      theme(panel.spacing = unit(0, "lines"),
            axis.text.y = element_text(size = 3),
            axis.text.x = element_text(angle = 270, hjust = 0, vjust = 0.5)) +
      scale_fill_manual(values = c('loss' = scales::muted('blue'),
                                   'dupli' = scales::muted('red'))) +
      scale_alpha(range = c(0, 1)) +
      facet_grid(annotation ~ event)
  }

plot_gains_and_losses(cnvs, 'cell_type')
plot_gains_and_losses(cnvs, 'cluster_name')
```

```{r plot_umaps, fig.height = 15}
list(
  # plot celltype
  cnvs %>%
    dplyr::filter(measure == 'proportion', event == 'dupli', chr == 'chr7') %>%
    ggplot(aes(x = UMAP1, y = UMAP2, colour = cell_type)) +
    geom_point(size = 0.05) +
    theme_classic() +
    coord_fixed(),
  # plot chr7 gain on umap
  cnvs %>%
    dplyr::filter(measure == 'proportion', event == 'dupli', chr == 'chr7') %>%
    dplyr::arrange(value) %>%
    ggplot(aes(x = UMAP1, y = UMAP2, colour = value)) +
    geom_point(size = 0.05) +
    theme_classic() +
    scale_colour_viridis_c() +
    coord_fixed() +
    ggtitle('proportion of genes duplicated on chr7')
) %>%
  patchwork::wrap_plots(ncol = 1)
```

#### chr7 gain

We are most interested in the chr7 gain events in fibroblasts. 

We will define cells with chr7 gain as those affecting >90% of genes.

```{r chr7_prop_dist}
p_dat <-
  cnvs %>%
  dplyr::filter(measure == 'proportion', event == 'dupli', chr == 'chr7') %>%
  dplyr::group_by(id, cluster_name, cell_type) %>%
  dplyr::mutate(median_proportion = median(value)) %>%
  dplyr::ungroup()

p_dat %>%
  ggplot(aes(x = value)) + 
  geom_density() +
  geom_vline(data = p_dat, aes(xintercept = median_proportion)) +
  ggh4x::facet_nested(~ subcluster + cell_type)

plot_prop_chr7_gain <- function(p_dat, position_arg) {
  p_dat %>%
    ggplot() +
    geom_bar(aes(x = cell_type, fill = value, group = value),
             position = position_arg) +
    theme_classic() +
    scale_fill_viridis_b(n.breaks = 9) 
}

plot_prop_chr7_gain(p_dat, position_stack(reverse = T))
plot_prop_chr7_gain(p_dat, position_fill(reverse = T))
p_dat %>%
  {split(., .$id)} %>%
  purrr::map2(names(.), ., function(id, df) {
    plot_prop_chr7_gain(df, position_fill(reverse = T)) +
      ggtitle(paste0('chr7 gain count - ', id))
    plot_prop_chr7_gain(df, position_stack(reverse = T)) +
      ggtitle(paste0('chr7 gain proportion - ', id))
  })
```

##### Pseudobulk

We pseudobulk chr7+ cells and perform differential expression analysis. In order to do this, we use the `glmGamPoi` package in R. This analysis follows [this tutorial](https://bioconductor.org/packages/devel/bioc/vignettes/glmGamPoi/inst/doc/pseudobulk.html).

```{r pseudobulk}
pseudobulk <- xfun::cache_rds({
  
  # pseudobulk the sce object by chr7 status
  reduced_sce <-
    glmGamPoi::pseudobulk(
      sce, 
      group_by = vars()
    )
  
}, file = 'pseudobulk.rds')
```
