---
title: "Gains and losses in rheumatoid arthritis data from inferCNV"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: false
    code_folding: hide
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
params:
  infercnv_dir: NULL
  cache_dir: NULL
  rerun: true
---

```{r setup, include = F, message = F, warning = F, class.source = 'fold-hide'}
# wd <- ifelse(Sys.info()['nodename'] == "mib119104s", '~/Volumes/', '') |> paste0('/lustre/scratch125/casm/team268im/at31/RA_som_mut/infercnv/') ; setwd(wd)
# rmarkdown::render('reports/infercnv.Rmd', output_file = 'driver_coverage.html', output_dir = 'reports/', params = list(rerun = T))

# libraries
library(magrittr)
library(ggplot2)
library(gridExtra)
library(Matrix)
library(SummarizedExperiment)
library(SingleCellExperiment)
library(glmGamPoi)

# dev.off() if there is a device
dev_off_if <- function() {if (length(dev.list()) != 0) { dev.off() }}

# chunk options
knitr::opts_knit$set(root.dir = '../')
knitr::opts_chunk$set(
  warning = FALSE, 
  dpi = 300, 
  message = FALSE,
  cache.path = params$cache_dir) 
```

inferCNV was run on the scRNA-seq data to call copy number variants in synovial cells. No reference group was used, so the average profile of all cells was used in each patient as the baseline. 

First, we load the CITE-seq object. We convert it to a SingleCellExperiment object. Next, we add the CNV metadata from the inferCNV run to the object.

```{r load_obj, results = 'asis'}
# dirs
data_dir <- '/lustre/scratch126/casm/team268im/mp34/scRNAseq_data/RA_ZhangEtal2023/processed_output/'
infercnv_dir <- 'out/Zhang2023/by_celltype/'

# convert to sce
sce <- xfun::cache_rds({
  
  # load object
  obj <- 
    readRDS(file.path(data_dir, 'all_cells_reference.rds'))
  
  # load rna data
  assays <-
    list(
      counts = 
        readRDS(file.path(data_dir, 'raw_mRNA_count_matrix.rds')),
      logcounts = 
        readRDS(file.path(data_dir, 'qc_mRNA_314011cells_log_normalized_matrix.rds')))
  
  # get cell order from counts
  cell_order <- colnames(assays$logcounts)
  
  # filter and reorder to cells that pass qc from counts
  assays$counts <- 
    assays$counts[rownames(assays$logcounts), 
                  cell_order]
  
  # generate coldata
  col_data <-
    obj$meta_data[cell_order, ]
  
  # populate sce
  sce <- 
    SingleCellExperiment::SingleCellExperiment(
      assays = assays,
      colData = col_data
    )
  
  # name, reorder, and add umap embeddings
  umap <- obj$umap$embedding
  rownames(umap) <- rownames(obj$meta_data)
  umap <- umap[cell_order, ]
  reducedDim(sce, 'UMAP_uwot') <- umap
  
  # get cnv metadata
  cnvs <- 
    list.files(
      infercnv_dir,
      pattern = 'map_metadata_from_infercnv.txt',
      recursive = T) %>%
    purrr::set_names(., dirname(.)) %>%
    purrr::map(function(file) {
      file.path(infercnv_dir, file) %>% 
        read.table(sep = '\t')
      }) %>%
    dplyr::bind_rows() 
  
  # left join cnvs to coldata, reorder, restore rownames
  col_data <- merge(col_data, cnvs, by = 'row.names', all.x = T)
  rownames(col_data) <- col_data[, 1]
  col_data <- 
    col_data[colnames(sce), -1] %>%
    DataFrame()
  
  # add to sce
  sce@colData <- col_data
  
  # return
  sce

}, file = 'sce.rds', rerun = params$rerun)
```

The `inferCNV` output file `map_metadata_from_infercnv.txt` contains columns with different CNV measures at whole-chromosome level per cell, with rownames denoting cells. The columns have the following values...

```{r echo = F, results = 'asis'}
tibble::tribble(
  ~name, ~`type - value`,
  '`has_*`', 'binary - whether any CNV / loss / duplication is found within the given chr',
  '`proportion_*`', 'proportion - proportion of genes that are part of any CNV / loss / duplication within the given chr',
  '`proportion_scaled_*`', 'proportion - same as proportion_*, but taking into account whether genes that are duplicated / lost are a single copy or two copies, as a weight',
  '`top_loss_*`', 'the top CNVs that affect the most genes'
) %>%
  knitr::kable()
```

First, we plot UMAPs by ID and celltype.

```{r plot_umaps, fig.height = 10}
dittoSeq::dittoDimPlot(sce, 'cell_type', size = 0.2) +
  coord_fixed()
```

Then, we visualise chr7 gains on the UMAP.

```{r plot_chr_gains}
dittoSeq::dittoDimPlot(
  sce, 'proportion_dupli_chr7', size = 0.1) +
  coord_fixed()
```

Now, we will plot gain and loss events by chromosome.

First, we plot single cell-level CNVs by direction and by celltype.

```{r plot_gains_and_losses_by_ct, fig.height = 20, eval = F}
# at single-cell level
unique(sce$cell_type) %>%
  purrr::walk(function(ct) {
    c('dupli', 'loss') %>%
      purrr::walk(function(direction) {
        p_dat <-
          sce@colData[!is.na(sce$proportion_cnv_chr1) & sce$cell_type == ct, ]
        p_mat <-
          p_dat[, grepl(paste0('proportion_', direction), colnames(p_dat))]
        p_annot <- 
          p_dat[, c('sample'), drop = F] %>%
          as.data.frame()
        p <-
          pheatmap::pheatmap(
            p_mat, 
            show_rownames = F, 
            annotation_row = p_annot,
            # hclust error: size cannot be NA nor exceed 65536
            cluster_rows = ifelse(nrow(p_mat) > 65536, F, T),
            legend = F,
            main = paste0('celltype: ', ct, '\ndirection: ', direction))
        print(p)
      })
  })
```

We also plot sample-x-celltype level events (mean-summarised). 

```{r plot_gains_and_losses_by_id}
# mean proportion loss / gain per chromosome per celltype per id
p_dat <-
  sce@colData %>%
  tibble::as_tibble() %>%
  dplyr::group_by(sample, cell_type) %>%
  dplyr::filter(!is.na(proportion_cnv_chr1)) %>%
  dplyr::summarise(
    dplyr::across(dplyr::starts_with('proportion_'), ~ mean(.x, na.rm = T))) %>%
  dplyr::ungroup() %>%
  dplyr::mutate(sample_x_ct = paste0(sample, '_', cell_type)) %>%
  tibble::column_to_rownames('sample_x_ct') 

# split gain and loss
p_dat <-
  c('dupli', 'loss') %>% 
  purrr::map(function(direction) {
    p_mat <-
      p_dat[, grepl(paste0('proportion_', direction), colnames(p_dat))]
    p_annot <- 
      p_dat[, c('sample', 'cell_type')] %>%
      as.data.frame()
    pheatmap::pheatmap(
      p_mat, 
      show_rownames = F, 
      annotation_row = p_annot,
      main = paste0('direction: ', direction))
  })
```

#### chr7 gain

We are most interested in the chr7 gain events in fibroblasts. We visualise their distribution across the dataset. 

```{r chr7_prop_dist}
# prep data
p_dat <-
  sce@colData %>%
  tibble::as_tibble() %>%
  dplyr::filter(!is.na(proportion_dupli_chr7)) %>%
  dplyr::group_by(sample) %>%
  dplyr::mutate(n_cells = dplyr::n())

# define functions
plot_prop_chr7_gain <- function(p_dat, title) {
  plot_func <- function(p_dat, position_arg) {
    p_dat %>%
      ggplot() +
      geom_bar(aes(x = cell_type, 
                   fill = proportion_dupli_chr7, 
                   group = proportion_dupli_chr7),
               position = position_arg) +
      scale_x_discrete(guide = guide_axis(angle = -45)) +
      theme_classic() +
      scale_fill_viridis_b(n.breaks = 9, limits = c(0, 1)) 
  }
  
  # plot both count (stack) and proportion (fill), annotate n cells
  list(
    plot_func(p_dat, position_fill(reverse = T)) +
      ggtitle(title) +
      theme(axis.text.x = element_blank(),
            axis.title.x = element_blank()) +
      ylab('count'),
    plot_func(p_dat, position_stack(reverse = T)) +
      ylab('proportion')) %>%
    patchwork::wrap_plots(ncol = 1) 
}

# plot, across all and by patient
plot_prop_chr7_gain(p_dat, 'chr7 gain - all cells')
p_dat %>%
  {split(., .$sample)} %>%
  purrr::walk2(names(.), ., function(id, df) {
    df %>%
    plot_prop_chr7_gain(
      paste0('chr7 gains - ', id, ' (', unique(df$n_cells), ' cells)')) %>%
      print()
  })
```

##### Pseudobulk

We pseudobulk chr7+ cells and perform differential expression analysis. In order to do this, we use the `glmGamPoi` package in R. This analysis follows a [tutorial](https://bioconductor.org/packages/devel/bioc/vignettes/glmGamPoi/inst/doc/pseudobulk.html) and the [glmGamPoi GitHub](https://github.com/const-ae/glmGamPoi).

We will define cells with chr7 gain as those affecting >90% of genes. We will pseudobulk by patient, celltype, and chr7 ploidy to produce a reduced object.

```{r pseudobulk_chr7_gain}
reduced_sce <- xfun::cache_rds({
  
  # we assign chr7 status as >90% of genes affected by a duplication
  sce@colData$chr7_gain <-
    as.numeric(sce@colData$proportion_dupli_chr7 > 0.9)
  
  # pseudobulk the sce object by patient and chr7 ploidy
  # subset to stromal cells
  reduced_sce <-
    glmGamPoi::pseudobulk(
      sce[, sce$cell_type == 'Stromal cell'], 
      group_by = vars(sample, chr7_gain)
    )
  
  # remove NA's
  reduced_sce <-
    reduced_sce[, !is.na(reduced_sce$sample) & !is.na(reduced_sce$chr7_gain)]
  
  # return
  reduced_sce
  
}, file = 'reduced_sce.rds', rerun = params$rerun)
```

With the reduced data, we can conduct differential expression analysis, treating the data like bulk RNA-seq data. 

We will identify which genes in stromal cells are changed most by chr7 gain. We do this by fitting one GLM model for each gene, with the chr7 status and the patient as factors.

```{r deseq, message = T}
# fit one GLM model for each gene
fit <- xfun::cache_rds({
  fit <-
  glmGamPoi::glm_gp(
    reduced_sce,
    design = ~ chr7_gain + sample,
    verbose = T
  )
  fit
}, file = 'fit.rds', rerun = params$rerun)

# show model output
summary(fit)
```

Next, we test for differential gene expression.

```{r test_ge}
# get differential expression
de_res <-
  glmGamPoi::test_de(
    fit, 
    contrast = 
      cond(chr7_gain = 1) -
      cond(chr7_gain = 0)
  )
```

```{r plot_de}
# minimum accepted adj_pval
min_p <- 0.05

# define up/downregulation
de_res <-
  de_res %>% 
  tibble::as_tibble() %>%
  dplyr::mutate(
    direction = dplyr::case_when(
      lfc < -1.5 & adj_pval < min_p ~ 'downregulated',
      lfc > 1.5 & adj_pval < min_p ~ 'upregulated',
      TRUE ~ NA
    )
  )

# plot
de_res %>%
  ggplot(aes(x = lfc, y = -log10(adj_pval), colour = direction)) +
  geom_point(size = 0.5) +
  geom_hline(yintercept = -log10(min_p), linetype = 'dashed') +
  geom_vline(xintercept = c(-1.5, 1.5), linetype = 'dashed') +
  ggrepel::geom_text_repel(
    data = 
      de_res %>% 
      dplyr::filter(!is.na(direction)) %>%
      dplyr::group_by(direction) %>%
      dplyr::filter(rank(adj_pval) <= 10 | rank(lfc, ties.method = 'min') <= 10),
    aes(label = name),
    min.segment.length = 0, nudge_x = 0.5) +
  theme_classic() +
  scale_colour_manual(values = c('upregulated' = 'darkred',
                                 'downregulated' = 'darkblue'), 
                      na.value = 'darkgrey')
```

These are the most strongly downregulated genes:

```{r get_downreg_degs}
de_res %>%
  tibble::as_tibble() %>% 
  dplyr::filter(adj_pval < 0.01) %>%
  dplyr::arrange(-lfc, adj_pval) %>%
  head(10) %>%
  knitr::kable()
```

These are the most strongly upregulated genes:

```{r get_upreg_degs}
de_res %>%
  tibble::as_tibble() %>% 
  dplyr::filter(adj_pval < 0.01) %>%
  dplyr::arrange(lfc, adj_pval) %>%
  head(10) %>%
  knitr::kable()
```
