---
title: "Creating figures for the paper"
author: "Alexandra Tidd"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    fig_width: 8
    keep_md: true
    code_folding: hide
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    theme: lumen
params:
  sce: NULL
  cell_group_name: NULL
  infercnv_dir: NULL
  metadata: NULL
  cache_dir: NULL
  rerun: true
  max_prop_for_wt: 0.5
  min_prop_for_cnv: 0.8
---

```{r setup, include = F, message = F, warning = F, class.source = "fold-hide"}
# module load R/4.4.0 ; Rscript -e 'rmarkdown::render("reports/paper_figures.Rmd", output_dir = "out/paper_figures/")' # nolint: line_length_linter.

# libraries
library(magrittr)
library(ggplot2)
library(SummarizedExperiment)
library(SingleCellExperiment)

# chunk options
knitr::opts_knit$set(root.dir = "../")
knitr::opts_chunk$set(warning = FALSE, dpi = 300, message = FALSE)
```

# Loading the inferCNV outputs

inferCNV was run on the scRNA-seq data to call copy number variants in synovial 
cells. No reference group was used, so the average profile of all cells was used 
in each patient as the baseline. 

First, we load the CITE-seq data as a `SingleCellExperiment` object. The CNV 
metadata from the `inferCNV` run has been aded to this object.

```{r load_sce, results = "asis", class.source = "fold-show"}
# load sce with gains
sce <-
  "out/Zhang2023/by_fibroblast_cluster/summary/gains_and_losses_cache/" %>%
  list.files(pattern = "sce_w_gains", full.names = TRUE) %>%
  readRDS()
```

# Quantifying CNVs of interest in the samples' fibroblast populations

We first prepare the data so that we can view the proportions of chr7 and chr5 
gains in each patient's fibroblast population.

```{r}
events <-
  sce@colData %>%
  tibble::as_tibble() %>%
  dplyr::select(sample, cell, cell_group, chr7_gain, chr5_gain) %>%
  tidyr::pivot_longer(-c("sample", "cell", "cell_group")) %>%
  dplyr::group_by(sample, cell, cell_group) %>%
  # remove cells in which no calls made
  dplyr::filter(!all(is.na(value))) %>%
  dplyr::mutate(event = dplyr::case_when(all(is.na(value) | value == 0) ~ "WT",
                                         value == 1 ~ name, TRUE ~ NA)) %>%
  dplyr::summarise(
    events = paste(sort(unique(na.omit(event))), collapse = ",") %>%
      factor(levels = c("chr5_gain,chr7_gain", "chr7_gain", "chr5_gain", "WT")),
    n_events = sum(value, na.rm = TRUE)) %>%
  # get props and counts
  dplyr::ungroup() %>%
  dplyr::add_count(sample, name = "n_cells_per_sample") %>%
  dplyr::add_count(sample, events, name = "n_cells_per_sample_x_event") %>%
  dplyr::distinct(sample, events,
                  n_cells_per_sample_x_event, n_cells_per_sample) %>%
  dplyr::mutate(prop_cells_per_sample_x_event = n_cells_per_sample_x_event /
                  n_cells_per_sample) %>%
  # refactor samples by prop non-WT cells
  dplyr::group_by(sample) %>%
  dplyr::mutate(
    n_wt_per_sample = dplyr::case_when(events == "WT" ~
                                         n_cells_per_sample_x_event,
                                       TRUE ~ 0),
    prop_non_wt_per_sample = (n_cells_per_sample - sum(n_wt_per_sample)) /
                  n_cells_per_sample) %>%
  dplyr::ungroup() %>%
  dplyr::mutate(sample = forcats::fct_reorder2(sample, n_cells_per_sample,
                                               prop_non_wt_per_sample)) %>%
  tidyr::pivot_longer(c("n_cells_per_sample_x_event",
                        "prop_cells_per_sample_x_event"))
```

# Filtering out patients with few fibroblasts

```{r}
min_n_fibroblasts_per_sample <- 100
```

We set the minimum number of fibroblasts per patient to 
`r min_n_fibroblasts_per_sample`. We can review how many patients pass the 
cut-off to give us reliable results.

```{r plot_fibroblast_dist}
n_fibroblasts <-
  sce@colData %>%
  tibble::as_tibble() %>%
  dplyr::count(sample)
n_fibroblasts %>%
  ggplot(aes(x = n)) +
  geom_histogram(binwidth = 20) +
  geom_vline(xintercept = min_n_fibroblasts_per_sample, linetype = "dashed", 
             colour = "red") +
  theme_classic()
```

```{r cut_off}
enough_fibroblasts <-
  dplyr::filter(n_fibroblasts, n >= min_n_fibroblasts_per_sample)$sample
n_fibroblasts %>%
  dplyr::summarise(`n >= 100` = sum(n >= 100),
                   `n < 100` = sum(n < 100),
                   n = dplyr::n()) %>%
  knitr::kable()
```

This leaves `r length(enough_fibroblasts)` / `r nrow(n_fibroblasts)` of the 
samples. We will filter the `SingleCellExperiment` object accordingly.

We can also review the relationship between the number of cells in each 
patient's fibroblast population and the proportions / counts of the CNVs of 
interest (chr7 gain and chr5 gain).

```{r plot_event_vs_fibroblast}
events %>%
  ggplot(aes(x = sample, y = value, fill = events)) +
  geom_col() +
  scale_x_discrete(guide = guide_axis(angle = -90)) +
  facet_wrap(~ name, scale = "free_y", ncol = 1) +
  theme_classic() +
  scale_fill_brewer(palette = "Dark2") +
  theme(legend.position = "bottom")
```

We can now apply the cut-off.

```{r apply_cutoff}
sce <- sce[, sce$sample %in% enough_fibroblasts]
events <- events %>% dplyr::filter(sample %in% enough_fibroblasts)
```

After applying the cut-off, we look for sample-size effects once again.

```{r plot_event_vs_fibroblast_after_cutoff}
events %>%
  ggplot(aes(x = sample, y = value, fill = events)) +
  geom_col() +
  scale_x_discrete(guide = guide_axis(angle = -90)) +
  facet_wrap(~ name, scale = "free_y", ncol = 1) +
  theme_classic() +
  scale_fill_brewer(palette = "Dark2") +
  theme(legend.position = "bottom")
```

We can see that there is still a strong relationship between the size of the
fibroblast populations by patient and the proportion of those cells that are
called as having a chr7 or chr5 gain. This is most likely related to with the
fact that we set `ref_group_names = NULL`. This compounds the effect of the
number of cells on the ability to detect CNVs (the false negative rate), 
because...

1. having fewer cells makes it less likely to capture these somewhat rare events

2. having fewer cells makes the reference profile noisier and less
representative of the 'baseline' expression levels across chromosomes, further
limiting CNV calling

Keeping this in mind, we will proceed.

# Plot 1 - chr7, chr5, and chr3 barplots by sample

```{r}
p <-
  sce@colData %>%
  tibble::as_tibble() %>%
  dplyr::add_count(sample) %>%
  tidyr::pivot_longer(cols = c("chr7_gain", "chr5_gain", "chr3_gain")) %>%
  dplyr::filter(!is.na(value)) %>%
  dplyr::mutate(value = as.character(value)) %>%
  ggplot(aes(x = reorder(sample, -n), fill = value)) +
  scale_x_discrete(guide = guide_axis(angle = -90)) +
  theme_classic() +
  scale_fill_brewer(palette = "Dark2") +
  facet_grid(name ~ .)
p + geom_bar() + labs(x = "sample", y = "n cells")
p + geom_bar(position = "fill") + labs(x = "sample", y = "proportion of cells")
```

# Plot 2 - chr7, chr5, and chr3 barplots by celltype

```{r}
cnvs <-
  "out/Zhang2023/by_celltype/summary/gains_and_losses_cache/" %>%
  list.files(pattern = "cnvs", full.names = TRUE) %>%
  readRDS()
p <-
  cnvs %>%
  dplyr::filter(event == "dupli", measure == "proportion",
                (value < 0.5 | value > 0.8),
                chr %in% c("chr7", "chr5", "chr3")) %>%
  dplyr::mutate(
    value = dplyr::case_when(value <= 0.5 ~ 0, value >= 0.8 ~ 1) %>%
      as.character()) %>%
  ggplot(aes(x = cell_type, fill = value)) +
  theme_classic() +
  facet_grid(chr ~ .) +
  scale_fill_brewer(palette = "Dark2") +
  scale_x_discrete(guide = guide_axis(angle = -45))
p + geom_bar()
p + geom_bar(position = "fill") + ylab("proportion")
```

# Plot 3 - number of transcripts per cell vs CNVs

Next, we look at the distribution of the the number of transcripts in cells in 
which chr7 gain has been detected versus those without. If the distribution is 
greatly skewed in one or the other, this would suggest that this technical 
variate is strongly affecting the performance of `inferCNV` between cells.

```{r}
colSums(sce@assays@data$counts) %>%
  tibble::enframe(name = "cell", value = "n_transcripts") %>%
  dplyr::inner_join(sce@colData %>% tibble::as_tibble()) %>%
  dplyr::filter(!is.na(chr7_gain)) %>%
  dplyr::mutate(chr7_gain = as.character(chr7_gain)) %>%
  ggplot(aes(x = n_transcripts, fill = chr7_gain)) +
  geom_histogram(alpha = 0.5) +
  theme_classic()
```

The distributions appear to be sensible. 

# Plot 4 - inferCNV plot with all individuals together


